{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "_Neural Data Science_\n",
    "\n",
    "Lecturer: Dr. Jan Lause, Prof. Dr. Philipp Berens\n",
    "\n",
    "Tutors: Jonas Beck, Fabio Seel, Julius Würzler\n",
    "\n",
    "Summer term 2025\n",
    "\n",
    "Student names: Nina Lutz, Mathis Nommensen\n",
    "\n",
    "LLM Disclaimer: Copilot was used to speed up the coding process in VS. ChatGPT was used for code review, nicer plots and debugging. ChatGPT was also used to aggregate information from papers on good practice and also used for quick lookups on biological information regarding genes and their functions.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Project 3: Single-cell data analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# %matplotlib notebook --> had to change this to compile\n",
    "%matplotlib inline\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib\n",
    "import string\n",
    "\n",
    "import scipy as sp\n",
    "from scipy import sparse\n",
    "import sklearn\n",
    "\n",
    "## add your packages ##\n",
    "\n",
    "import time\n",
    "import pickle\n",
    "import memory_profiler\n",
    "import seaborn as sns\n",
    "import scipy.stats as stats\n",
    "import umap.umap_ as umap\n",
    "import igraph as ig\n",
    "import leidenalg\n",
    "#import umap\n",
    "\n",
    "from sklearn.decomposition import PCA\n",
    "from sklearn.manifold import TSNE\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.preprocessing import LabelEncoder\n",
    "from sklearn.neighbors import NearestNeighbors\n",
    "from sklearn.metrics import adjusted_rand_score\n",
    "from sklearn.mixture import GaussianMixture\n",
    "from sklearn.cluster import KMeans\n",
    "from sklearn.neighbors import KNeighborsClassifier\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.metrics import recall_score\n",
    "from sklearn.cluster import KMeans\n",
    "from sklearn.metrics import silhouette_score, adjusted_mutual_info_score\n",
    "from sklearn.feature_selection import VarianceThreshold\n",
    "from sklearn.cross_decomposition import CCA\n",
    "from sklearn.ensemble import RandomForestRegressor\n",
    "from sklearn.linear_model import LinearRegression, ElasticNet\n",
    "from sklearn.model_selection import train_test_split, cross_validate\n",
    "from sklearn.metrics import r2_score, mean_squared_error\n",
    "from sklearn.pipeline import make_pipeline\n",
    "\n",
    "from scipy.stats import pearsonr\n",
    "from scipy.stats import nbinom\n",
    "\n",
    "from statsmodels.stats.multitest import fdrcorrection\n",
    "\n",
    "from gprofiler import GProfiler\n",
    "\n",
    "\n",
    "%load_ext memory_profiler\n",
    "\n",
    "from pathlib import Path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import black\n",
    "import jupyter_black\n",
    "\n",
    "jupyter_black.load(line_length=79)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "variables_path = Path(\"../results/variables\")\n",
    "figures_path = Path(\"../results/figures\")\n",
    "data_path = Path(\"../data\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plt.style.use(\"matplotlib_style.txt\")\n",
    "plt.style.use(\n",
    "    \"../matplotlib_style.txt\"\n",
    ")  # had to change this as well to succesfully compile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.random.seed(42)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Project and data description\n",
    "\n",
    "In this project, we are going to work with the typical methods and pipelines used in single-cell data analysis and get some hands-on experience with the techniques used in the field. For that, we will be using Patch-seq multimodal data from cortical neurons in mice, from Scala et al. 2021 (https://www.nature.com/articles/s41586-020-2907-3#Sec7). From the different data modalities they used, we will focus on transcriptomics and electrophysiological data. \n",
    "\n",
    "In a real-world scenario, single cell data rarely comes with any \"ground truth\" labels. Often, the goal of researchers after measuring cells is to precisely classify them, grouping them into families or assigning them cell types based on the recorded features. This is normally done using usupervised methods, such as clustering methods.\n",
    "\n",
    "However, the single-cell data that we are using in this project has some cell types assigned to each cell. These are not \"ground truth\" type annotations, but were one of the results from the original Scala et al. work. Still, we are going to use those annotations for validation (despite them not really being ground truth) to sanity-check some of our analyses, such as visualizations, clustering, etc. We will mainly work with cell types (`rna_types`, 77 unique types) and cell families (`rna_families`, 9 unique families).\n",
    "\n",
    "From the transcriptomics mRNA counts, we will only work with the exon counts for simplicity. Some of the electrophysiological features are not high-quality recordings, therefore we will also filter them out."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import data"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Meta data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# META DATA\n",
    "\n",
    "meta = pd.read_csv(data_path / \"m1_patchseq_meta_data.csv\", sep=\"\\t\")\n",
    "\n",
    "cells = meta[\"Cell\"].values\n",
    "\n",
    "layers = meta[\"Targeted layer\"].values.astype(\"str\")\n",
    "cre = meta[\"Cre\"].values\n",
    "yields = meta[\"Yield (pg/µl)\"].values\n",
    "yields[yields == \"?\"] = np.nan\n",
    "yields = yields.astype(\"float\")\n",
    "depth = meta[\"Soma depth (µm)\"].values\n",
    "depth[depth == \"Slice Lost\"] = np.nan\n",
    "depth = depth.astype(float)\n",
    "thickness = meta[\"Cortical thickness (µm)\"].values\n",
    "thickness[thickness == 0] = np.nan\n",
    "thickness = thickness.astype(float)\n",
    "traced = meta[\"Traced\"].values == \"y\"\n",
    "exclude = meta[\"Exclusion reasons\"].values.astype(str)\n",
    "exclude[exclude == \"nan\"] = \"\"\n",
    "\n",
    "mice_names = meta[\"Mouse\"].values\n",
    "mice_ages = meta[\"Mouse age\"].values\n",
    "mice_cres = np.array(\n",
    "    [\n",
    "        c if c[-1] != \"+\" and c[-1] != \"-\" else c[:-1]\n",
    "        for c in meta[\"Cre\"].values\n",
    "    ]\n",
    ")\n",
    "mice_ages = dict(zip(mice_names, mice_ages))\n",
    "mice_cres = dict(zip(mice_names, mice_cres))\n",
    "\n",
    "print(\"Number of cells with measured depth:    \", np.sum(~np.isnan(depth)))\n",
    "print(\"Number of cells with measured thickness:\", np.sum(~np.isnan(thickness)))\n",
    "print(\"Number of reconstructed cells:          \", np.sum(traced))\n",
    "\n",
    "sliceids = meta[\"Slice\"].values\n",
    "a, b = np.unique(sliceids, return_counts=True)\n",
    "assert np.all(b <= 2)\n",
    "print(\"Number of slices with two cells:        \", np.sum(b == 2))\n",
    "\n",
    "# Some consistency checks\n",
    "assert np.all(\n",
    "    [\n",
    "        np.unique(meta[\"Date\"].values[mice_names == m]).size == 1\n",
    "        for m in mice_names\n",
    "    ]\n",
    ")\n",
    "assert np.all(\n",
    "    [\n",
    "        np.unique(meta[\"Mouse age\"].values[mice_names == m]).size == 1\n",
    "        for m in mice_names\n",
    "    ]\n",
    ")\n",
    "assert np.all(\n",
    "    [\n",
    "        np.unique(meta[\"Mouse gender\"].values[mice_names == m]).size == 1\n",
    "        for m in mice_names\n",
    "    ]\n",
    ")\n",
    "assert np.all(\n",
    "    [\n",
    "        np.unique(meta[\"Mouse genotype\"].values[mice_names == m]).size == 1\n",
    "        for m in mice_names\n",
    "    ]\n",
    ")\n",
    "assert np.all(\n",
    "    [\n",
    "        np.unique(meta[\"Mouse\"].values[sliceids == s]).size == 1\n",
    "        for s in sliceids\n",
    "    ]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "meta.columns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### \"Ground truth labels\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# filter out low quality cells in term of RNA\n",
    "print(\n",
    "    \"There are\",\n",
    "    np.sum(meta[\"RNA family\"] == \"low quality\"),\n",
    "    \"cells with low quality RNA recordings.\",\n",
    ")\n",
    "exclude_low_quality = meta[\"RNA family\"] != \"low quality\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rna_family = meta[\"RNA family\"][exclude_low_quality]\n",
    "rna_type = meta[\"RNA type\"][exclude_low_quality]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(len(np.unique(rna_family)))\n",
    "print(len(np.unique(rna_type)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pickle_in = open(data_path / \"dict_rna_type_colors.pkl\", \"rb\")\n",
    "dict_rna_type_colors = pickle.load(pickle_in)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rna_type_colors = np.vectorize(dict_rna_type_colors.get)(rna_type)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Transcriptomic data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# READ COUNTS\n",
    "data_exons = pd.read_csv(\n",
    "    data_path / \"m1_patchseq_exon_counts.csv.gz\", na_filter=False, index_col=0\n",
    ")\n",
    "\n",
    "assert all(cells == data_exons.columns)\n",
    "genes = np.array(data_exons.index)\n",
    "\n",
    "# filter out low quality cells in term of rna family\n",
    "exonCounts = data_exons.values.transpose()[exclude_low_quality]\n",
    "print(\"Count matrix shape (exon):  \", exonCounts.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# GENE LENGTH\n",
    "\n",
    "data = pd.read_csv(data_path / \"gene_lengths.txt\")\n",
    "assert all(data[\"GeneID\"] == genes)\n",
    "exonLengths = data[\"exon_bp\"].values"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Electrophysiological features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# EPHYS DATA\n",
    "\n",
    "ephysData = pd.read_csv(data_path / \"m1_patchseq_ephys_features.csv\")\n",
    "ephysNames = np.array(ephysData.columns[1:]).astype(str)\n",
    "ephysCells = ephysData[\"cell id\"].values\n",
    "ephysData = ephysData.values[:, 1:].astype(\"float\")\n",
    "names2ephys = dict(zip(ephysCells, ephysData))\n",
    "ephysData = np.array(\n",
    "    [\n",
    "        names2ephys[c] if c in names2ephys else ephysData[0] * np.nan\n",
    "        for c in cells\n",
    "    ]\n",
    ")\n",
    "\n",
    "print(\"Number of cells with ephys data:\", np.sum(np.isin(cells, ephysCells)))\n",
    "\n",
    "assert np.sum(~np.isin(ephysCells, cells)) == 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filtering ephys data\n",
    "\n",
    "features_exclude = [\n",
    "    \"Afterdepolarization (mV)\",\n",
    "    \"AP Fano factor\",\n",
    "    \"ISI Fano factor\",\n",
    "    \"Latency @ +20pA current (ms)\",\n",
    "    \"Wildness\",\n",
    "    \"Spike frequency adaptation\",\n",
    "    \"Sag area (mV*s)\",\n",
    "    \"Sag time (s)\",\n",
    "    \"Burstiness\",\n",
    "    \"AP amplitude average adaptation index\",\n",
    "    \"ISI average adaptation index\",\n",
    "    \"Rebound number of APs\",\n",
    "]\n",
    "features_log = [\n",
    "    \"AP coefficient of variation\",\n",
    "    \"ISI coefficient of variation\",\n",
    "    \"ISI adaptation index\",\n",
    "    \"Latency (ms)\",\n",
    "]\n",
    "\n",
    "X = ephysData[exclude_low_quality]\n",
    "print(X.shape)\n",
    "for e in features_log:\n",
    "    X[:, ephysNames == e] = np.log(X[:, ephysNames == e])\n",
    "X = X[:, ~np.isin(ephysNames, features_exclude)]\n",
    "\n",
    "keepcells = ~np.isnan(np.sum(X, axis=1))\n",
    "X = X[keepcells, :]\n",
    "print(X.shape)\n",
    "\n",
    "X = X - X.mean(axis=0)\n",
    "ephysData_filtered = X / X.std(axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.sum(np.isnan(ephysData_filtered))"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Research questions to investigate\n",
    "\n",
    "**1) Inspect the data by computing key statistics.** For RNA counts, you can compute and plot statistics, e.g. total counts per cell, number of expressed genes per cell, mean count per gene, variance per gene, mean-variance relationship... See https://www.embopress.org/doi/full/10.15252/msb.20188746 for common quality control statistics. Keep in mind that the RNA data in this project is read counts, not UMI counts, so it is not supposed to follow a Poisson distribution.> To get an idea of the technical noise in the data, you can plot count distributions of single genes within cell types (like in the lecture). \n",
    "\n",
    "Similarly, you can compute and plot statistics over the electrophyiological data. Also, investigate the distribution of \"ground truth\" labels. Comment about other relevant metadata, and think if you can use it as some external validation for other analyses. If you do use other metadata throughout the project, explain why and what you get out of it. Take into account that certain features may not be very informative for our purposes (e.g. mouse age), so only choose features that provide you with useful information in this context. If you want to get additional information about the metadata, have a look at the extended data section in the original publication (e.g., cre-lines in Figure 1c in the extended data).\n",
    "\n",
    "**2) Normalize & transform the data; select genes & apply PCA.** There are several ways of normalizing the RNA count data (Raw, CPM, CPMedian, RPKM, see https://www.reneshbedre.com/blog/expression_units.html, https://translational-medicine.biomedcentral.com/articles/10.1186/s12967-021-02936-w). Take into account that there are certain normalizations that only make sense for UMI data, but not for this read count data. You also explored different transformations in the assignment (none, log, sqrt). Compare how the different transformations change the two-dimensional visualization. After normalization and transformation, choose a set of highly variable genes (as demonstrated in the lecture) and apply PCA. Play with the number of selected genes and the number of PCA components, and again compare their effects on the two-dimensional visualization.\n",
    "\n",
    "**3) Two-dimensional visualization.** To visualize the RNA count data after normalization, transformation, gene selection and PCA, try different methods (just PCA, t-SNE, UMAP, ..) and vary their parameters (exaggeration, perplexity, ..). Compare them using quantitative metrics (e.g., kNN accuracy in high-dim vs. two-dim, kNN recall). Please refer to Lause et al., 2024 (https://doi.org/10.1371/journal.pcbi.1012403) where many of these metrics are discussed and explained to make an informed choice on which metrics to use. Think about also using the electrophysiological features and other metadata to enhance different visualizations.\n",
    "\n",
    "**4) Clustering.** To find cell types in the RNA count data, you will need to look for clusters. Try different clustering methods (leiden, GMM). Implement a negative binomial mixture model. For that you can follow a similar method that what is described in Harris et al. 2018 (https://journals.plos.org/plosbiology/article?id=10.1371/journal.pbio.2006387#abstract0), with fixed r (r=2). Feel free to simplify the setup from the paper and not optimize over the set of important genes S but fix it instead, or skip the split and merge part of their clustering algorithm. A vanilla NBMM implementation should suffice. Take into account that the NBMM tries to cluster data that follows a negative binomial distribution. Therefore, it does not make sense to apply this clustering method to all kinds of normalized and transformed data. Please refer to the Harris et al. 2018 publication for the appropriate choice of normalization, and reflect on why this normalization makes sense. Evaluate your clustering results (metrics, compare number of clusters to original labels,...).\n",
    "\n",
    "**5) Correlation between electrophysiological features and genes/PCs.** Finally, connect RNA counts and functional data: Most likely, there will be interesting relationships between the transcriptomic and electrophyiological features in this data. Find these correlations and a way of visualizing them. In studying correlations using the PCA-reduced version of the transcriptomics data, it could be interesting to study PC loadings to see which genes are dominating which PCs. For other advanced analyses, you can get inspitation from Kobak et al., 2021 (https://doi.org/10.1111/rssc.12494).\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Task 1\n",
    "\n",
    "Outline of data exploration steps was mainly inspired by Luecken, M.D. & Theis, F.J. (2019)\n",
    "\n",
    "## 1.1 QC Statistics per cell\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# total counts per cell (count depth)\n",
    "\n",
    "# exonCounts  # shape = 1232 cells x 42466 genes\n",
    "total_counts_per_cell = exonCounts.sum(axis=1)\n",
    "\n",
    "# plot\n",
    "fig, axes = plt.subplots(1, 3, figsize=(15, 4))\n",
    "\n",
    "axes[0].hist(\n",
    "    total_counts_per_cell, bins=50, color=\"lightblue\", edgecolor=\"black\"\n",
    ")\n",
    "axes[0].set_xlabel(\"Total RNA counts per cell\")\n",
    "axes[0].set_ylabel(\"Number of cells\")\n",
    "axes[0].set_title(\"Distribution of Total Counts Per Cell\")\n",
    "\n",
    "axes[1].bar(\n",
    "    range(len(total_counts_per_cell)),\n",
    "    total_counts_per_cell,\n",
    "    width=1,\n",
    "    alpha=0.5,\n",
    "    color=\"blue\",\n",
    ")\n",
    "axes[1].set_yscale(\"log\")\n",
    "axes[1].set_xlabel(\"Cell index\")\n",
    "axes[1].set_ylabel(\"Total RNA counts\")\n",
    "axes[1].set_title(\"Total Counts Per Cell\")\n",
    "\n",
    "# Rank-ordered plot of total counts per cell - Figure 2C in the paper\n",
    "sorted_counts = np.sort(total_counts_per_cell)[::-1]\n",
    "\n",
    "axes[2].plot(sorted_counts, color=\"blue\")\n",
    "axes[2].set_yscale(\"log\")\n",
    "axes[2].set_xlabel(\"Ranked Cell Index\")\n",
    "axes[2].set_ylabel(\"Total Counts (log scale)\")\n",
    "axes[2].set_title(\"Rank-ordered Total Counts per Cell\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*Interpretation:*\n",
    "\n",
    "Left:\n",
    "* The histogram reveals a right-skewed distribution of total RNA counts per cell, with most cells clustered between 0 and 5 million counts.\n",
    "\n",
    "Middle:\n",
    "* shows that most cells have a comparable number of total counts, with a few outliers having significantly higher and lower counts (e.g. cells around index 400 and 500).\n",
    "\n",
    "Right:\n",
    "* The rank-ordered curve shows a gradual decay in RNA counts, with a steep drop-off in low-quality cells at the tail—typical of single-cell data \n",
    "* most cells have comparable total cells\n",
    "* the first ~20 cells have a lot more counts than the rest\n",
    "* the last ~100 cells have very few counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# number of expressed genes per cell\n",
    "expressed_genes_per_cell = (exonCounts > 0).sum(axis=1)\n",
    "\n",
    "# plot\n",
    "fig, axes = plt.subplots(1, 2, figsize=(10, 4))\n",
    "\n",
    "axes[0].scatter(range(len(expressed_genes_per_cell)), expressed_genes_per_cell)\n",
    "axes[0].set_xlabel(\"Cell Index\")\n",
    "axes[0].set_ylabel(\"Expressed Genes\")\n",
    "axes[0].set_title(\"Total Number of Expressed Genes per Cell\")\n",
    "\n",
    "axes[1].hist(expressed_genes_per_cell, bins=50)\n",
    "axes[1].set_xlabel(\"Total number of expressed genes per cell\")\n",
    "axes[1].set_ylabel(\"Number of cells\")\n",
    "axes[1].set_title(\"Distribution of Expressed Genes Per Cell\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*Interpretation:*\n",
    "\n",
    "Left: Total number of expressed genes per cell\n",
    "* The scatter plot shows considerable variation in the number of genes detected per cell, with a spread ranging from around 2,000 to over 17,000 genes, whereas most cells have between 5000 and 11,000 genes expressed.\n",
    "\n",
    "Right: Distribution of expressed genes per cell\n",
    "* Shows a similar pattern as the left plot, with most cells having between 2000 and 11,000 genes expressed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# fraction of mitochondrial genes\n",
    "\n",
    "# mt_like_genes = [g for g in genes if \"mt\" in g.lower()]\n",
    "# print(mt_like_genes)\n",
    "# after exploring the gene names, we can assume that mitochondrial genes start with \"mt-\" or \"MT-\"\n",
    "mt_gene_mask = np.char.startswith(genes.astype(str).astype(\"U\"), \"mt-\")\n",
    "\n",
    "print(\"Number of mitochondrial genes found:\", np.sum(mt_gene_mask))\n",
    "\n",
    "# Sum counts over mitochondrial genes per cell\n",
    "mt_counts_per_cell = exonCounts[:, mt_gene_mask].sum(axis=1)\n",
    "\n",
    "# Fraction mitochondrial\n",
    "fraction_mito = mt_counts_per_cell / total_counts_per_cell\n",
    "\n",
    "# Print statistics\n",
    "print(f\"Mean mitochondrial fraction: {np.mean(fraction_mito):.3f}\")\n",
    "print(\"Cells with >20% mitochondrial:\", np.sum(fraction_mito > 0.2))\n",
    "\n",
    "# Plot\n",
    "plt.figure(figsize=(8, 4))\n",
    "plt.hist(fraction_mito, bins=50, color=\"salmon\", edgecolor=\"black\")\n",
    "plt.xlabel(\"Fraction of Mitochondrial Counts per Cell\")\n",
    "plt.ylabel(\"Number of Cells\")\n",
    "plt.title(\"Distribution of Mitochondrial RNA Content\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*Interpretation:*\n",
    "\n",
    "* most cells have a low fraction of mitochondrial genes (~1.7%), with a few outliers having a higher fraction (e.g. cells around index 400 and 500)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1.2 QC Statistics per gene\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# mean expression across all cells\n",
    "mean_expression_across_cells = exonCounts.mean(axis=0)\n",
    "print(\"Mean expression across all cells:\", mean_expression_across_cells.shape)\n",
    "\n",
    "# variance across all cells\n",
    "variance_expression_across_cells = exonCounts.var(axis=0)\n",
    "print(\"Variance across all cells:\", variance_expression_across_cells.shape)\n",
    "\n",
    "# plot looked sparse, so log transform the data\n",
    "fig, ax = plt.subplots(figsize=(8, 6))\n",
    "ax.scatter(\n",
    "    mean_expression_across_cells + 1e-3,  # avoid log(0)\n",
    "    variance_expression_across_cells + 1e-3,\n",
    "    s=5,\n",
    "    alpha=0.5,\n",
    ")\n",
    "ax.set_xscale(\"log\")\n",
    "ax.set_yscale(\"log\")\n",
    "ax.set_xlabel(\"Mean Expression Across Cells (log scale)\")\n",
    "ax.set_ylabel(\"Variance Across Cells (log scale)\")\n",
    "ax.set_title(\"Mean vs Variance of Gene Expression (log scale)\")\n",
    "x_vals = np.linspace(\n",
    "    min(mean_expression_across_cells + 1e-3),\n",
    "    max(mean_expression_across_cells),\n",
    "    100,\n",
    ")\n",
    "ax.plot(x_vals, x_vals, color=\"red\", linestyle=\"--\", label=\"y = x\")\n",
    "ax.legend()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*Interpretation:*\n",
    "\n",
    "* The log-log plot reveals a clear mean-variance relationship, where genes with higher mean expression across cells also tend to have higher variance\n",
    "* Almost all genes lie above the y=x line (variance = mean), indicating that the variance is generally higher than the mean for most genes (overdispersion)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# dropout rate / fraction of cells where a gene has zero counts\n",
    "dropout_rate_per_gene = (exonCounts == 0).sum(axis=0) / exonCounts.shape[0]\n",
    "\n",
    "# Quick summary\n",
    "print(\n",
    "    f\"Mean dropout rate across all genes: {np.mean(dropout_rate_per_gene):.3f}\"\n",
    ")\n",
    "\n",
    "fig, axes = plt.subplots(1, 2, figsize=(14, 5))\n",
    "\n",
    "# Plot Histogram of dropout rates\n",
    "axes[0].hist(dropout_rate_per_gene, bins=50, color=\"green\", edgecolor=\"black\")\n",
    "axes[0].set_xlabel(\"Dropout Rate (Fraction of Cells with Zero Counts)\")\n",
    "axes[0].set_ylabel(\"Number of Genes\")\n",
    "axes[0].set_title(\"Dropout Rate per Gene\")\n",
    "\n",
    "# Scatter plot of dropout rate vs mean expression\n",
    "axes[1].scatter(\n",
    "    mean_expression_across_cells + 1e-3,  # avoid log(0)\n",
    "    dropout_rate_per_gene,\n",
    "    s=5,\n",
    "    alpha=0.5,\n",
    ")\n",
    "axes[1].set_xscale(\"log\")\n",
    "axes[1].set_ylim(0, 1.05)\n",
    "axes[1].set_xlabel(\"Mean Expression (log scale)\")\n",
    "axes[1].set_ylabel(\"Dropout Rate\")\n",
    "axes[1].set_title(\"Dropout Rate vs. Mean Expression\")\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*Interpretation:*\n",
    "\n",
    "Left:\n",
    "* the majority of genes have a very high dropout rate (mean ~83%), with a large number expressed in fewer than 15-20% of cells and a peak near 100% dropout, i.e. many genes are completely unexpressed in most cells\n",
    "\n",
    "Right:\n",
    "* Strong inverse relationship: Highly expressed genes are detected in nearly all cells (low dropout), while lowly expressed genes are often undetected (high dropout)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**count distributions of single genes within cell types**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# identify cell types\n",
    "unique_cell_types = np.unique(rna_type)\n",
    "\n",
    "# pick genes:\n",
    "# Sst: somatostatin, an inhibitory neuron marker\n",
    "# Slc17a7: an excitatory neuron marker (VGLUT1)\n",
    "# Snap25: a pan-neuronal gene\n",
    "genes_to_plot = [\"Sst\", \"Slc17a7\", \"Snap25\"]\n",
    "fig, axes = plt.subplots(nrows=3, ncols=1, figsize=(10, 8))\n",
    "\n",
    "for ax, gene_name in zip(axes, genes_to_plot):\n",
    "    gene_indices = np.where(genes == gene_name)[0]\n",
    "    gene_index = gene_indices[0]\n",
    "\n",
    "    # get gene counts\n",
    "    gene_counts = exonCounts[:, gene_index]\n",
    "\n",
    "    # group counts by cell type\n",
    "    data = []\n",
    "    for ct in unique_cell_types:\n",
    "        mask = np.array(rna_type) == ct\n",
    "        counts_in_group = gene_counts[mask]\n",
    "        data.append(counts_in_group)\n",
    "\n",
    "    ax.boxplot(\n",
    "        data, tick_labels=unique_cell_types, showfliers=False\n",
    "    )  # showfliers=False option removes extreme outliers\n",
    "    ax.set_title(f\"Expression of {gene_name}\")\n",
    "\n",
    "    ###M: added logscale to y-axis\n",
    "    ax.set_yscale(\"log\")  # log scale for better visibility\n",
    "\n",
    "    ax.set_ylabel(\"Counts\")\n",
    "    ax.set_xlabel(\"Cell Type\")\n",
    "    ax.tick_params(axis=\"x\", rotation=90, labelsize=8)\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*Interpretation:*\n",
    "\n",
    "Sst:\n",
    "* Expression is highly enriched in specific inhibitoy cell types (notably those starting with \"Sst\"), occasionally enriched expression in other celltypes, but no broader pattern visible.\n",
    "\n",
    "Slc17a7:\n",
    "* Expression is enriched in many cells, but most obviously in cell types from L2 to L6, which belong to a family of excitatory neurons. This suggests that Slc17a7 is a marker for excitatory neurons in the cortex, as it is known to be expressed in glutamatergic neurons.\n",
    "\n",
    "Snap25:\n",
    "* This gene shows widespread expression across nearly all cell types, reflecting its function as a general neuronal marker involved in synaptic transmission"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1.3 Statistics for electrophysiological features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# features overview\n",
    "\n",
    "# List of cleaned electrophysiological features retained for analysis\n",
    "ephysNames_filtered = ephysNames[\n",
    "    ~np.isin(ephysNames, features_exclude)\n",
    "]  # see above\n",
    "\n",
    "print(\n",
    "    \"Remaining electrophysiological features (n = {}) for analysis\".format(\n",
    "        len(ephysNames_filtered)\n",
    "    )\n",
    ")\n",
    "for i, name in enumerate(ephysNames_filtered, 1):\n",
    "    print(f\"{i:2d}. {name}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# descriptive statistics of ephysiological features (keep in mind that data is already standardized)\n",
    "\n",
    "# dictionary to collect stats\n",
    "stats_dict = {\n",
    "    \"Mean\": [],\n",
    "    \"Std\": [],\n",
    "    \"Min\": [],\n",
    "    \"Max\": [],\n",
    "    \"Median\": [],\n",
    "    \"Skewness\": [],\n",
    "}\n",
    "\n",
    "# Compute stats per feature\n",
    "for i in range(ephysData_filtered.shape[1]):\n",
    "    # data = X[:, i]\n",
    "    data = ephysData_filtered[:, i]\n",
    "\n",
    "    # Collect statistics\n",
    "    stats_dict[\"Mean\"].append(np.mean(data))\n",
    "    stats_dict[\"Std\"].append(np.std(data))\n",
    "    stats_dict[\"Min\"].append(np.min(data))\n",
    "    stats_dict[\"Max\"].append(np.max(data))\n",
    "    stats_dict[\"Median\"].append(np.median(data))\n",
    "    stats_dict[\"Skewness\"].append(stats.skew(data))\n",
    "\n",
    "# Convert to DataFrame\n",
    "feature_stats_df = pd.DataFrame(stats_dict, index=ephysNames_filtered)\n",
    "\n",
    "print(\"Basic statistics of electrophysiological features (standardized):\")\n",
    "display(feature_stats_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*Interpretation:*\n",
    "\n",
    "Overall:\n",
    "* mean and standard deviation are (roughly) 0 and 1, respectively, as the data is standardized\n",
    "* Min/Max values show range in z-score units\n",
    "\n",
    "Mentionable Observations:\n",
    "* ISI adaptation index (ske=2.48), Rebound (mV) (2.05), Sag ratio (+2.06), and Rheobase (pA) (+1.64) are strongly right-skew, indicating that most cells have low values, but a smaller subset show much higher values.\n",
    "* AP amplitude adaptation index (skew = -1.13) are left-skewed, indicating that most cells have high values (i.e. high adaptation), but a smaller subset show much lower values.\n",
    "* Membrane time constant, Input resistance, and Max number of APs have high max values (>3.8 z-score), indicating large variability among cells in how they respond to current input.\n",
    "* Several features show approximately symmetric distributions, like AP threshold (skew = −0.06), Afterhyperpolarization (mV) (skew = +0.03) and Resting membrane potential (skew = −0.12). This suggests that these properties are well-centered and consistent across cells.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot distributions of standardized electrophysiological features\n",
    "n_features = len(ephysNames_filtered)\n",
    "n_cols = 4\n",
    "n_rows = int(np.ceil(n_features / n_cols))\n",
    "\n",
    "fig, axes = plt.subplots(n_rows, n_cols, figsize=(n_cols * 4, n_rows * 3))\n",
    "\n",
    "axes = axes.flatten()\n",
    "\n",
    "for i, feature in enumerate(ephysNames_filtered):\n",
    "    ax = axes[i]\n",
    "    sns.histplot(\n",
    "        ephysData_filtered[:, i],\n",
    "        bins=30,\n",
    "        kde=True,\n",
    "        color=\"skyblue\",\n",
    "        stat=\"density\",\n",
    "        ax=ax,\n",
    "    )\n",
    "    sns.kdeplot(\n",
    "        ephysData_filtered[:, i],\n",
    "        color=\"darkblue\",\n",
    "        linewidth=1,\n",
    "        ax=ax,\n",
    "    )  # smoothed density plot (darkblue line)\n",
    "\n",
    "    ax.axvline(0, color=\"red\", linestyle=\"--\", linewidth=1)\n",
    "    ax.set_title(feature, fontsize=9, fontweight=\"bold\")\n",
    "    ax.set_xlabel(\"Standardized value\")\n",
    "    ax.set_ylabel(\"Density\")\n",
    "\n",
    "# Remove empty subplots\n",
    "for j in range(i + 1, len(axes)):\n",
    "    fig.delaxes(axes[j])\n",
    "\n",
    "fig.suptitle(\n",
    "    \"Distributions of Standardized Electrophysiological Features\",\n",
    "    fontsize=14,\n",
    "    fontweight=\"bold\",\n",
    ")\n",
    "plt.tight_layout(rect=[0, 0, 1, 0.97])\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*Interpretation:*\n",
    "* means (0) and standard deviations (1) are consistent with standardized data\n",
    "* skewness is clearly visible\n",
    "* generally, the distributions show clearly what the statistics suggest: all features are centered around zero with unit variance, and the observed skewness in the plots aligns with the computed skewness values, highlighting the diversity in distribution shapes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# DataFrame with standardized ephys data\n",
    "ephys_df = pd.DataFrame(ephysData_filtered, columns=ephysNames_filtered)\n",
    "\n",
    "# Compute Pearson correlation matrix\n",
    "corr_matrix = ephys_df.corr()\n",
    "\n",
    "# Test\n",
    "print(\"Correlation matrix shape:\", corr_matrix.shape)\n",
    "\n",
    "# \"high correlation\" threshold\n",
    "# commonly, a correlation |r| > 0.6 is considered \"strong\"\n",
    "threshold = 0.6\n",
    "high_corr_pairs = []\n",
    "\n",
    "# Iterate over the upper triangle of the correlation matrix (excluding the diagonal) because the matrix is symmetric\n",
    "for i in range(len(corr_matrix.columns)):\n",
    "    for j in range(i + 1, len(corr_matrix.columns)):\n",
    "        corr_value = corr_matrix.iloc[i, j]\n",
    "        if abs(corr_value) > threshold:\n",
    "            feature_i = corr_matrix.columns[i]\n",
    "            feature_j = corr_matrix.columns[j]\n",
    "            high_corr_pairs.append((feature_i, feature_j, corr_value))\n",
    "\n",
    "# Print results\n",
    "print(\n",
    "    \"Feature pairings with high correlations (|corr| > {:.2f}):\".format(\n",
    "        threshold\n",
    "    )\n",
    ")\n",
    "for feature1, feature2, corr in sorted(\n",
    "    high_corr_pairs, key=lambda x: -abs(x[2])\n",
    "):\n",
    "    print(f\"{feature1:30s} x {feature2:30s}: {corr:+.2f}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plotting the correlation matrix\n",
    "plt.figure(figsize=(12, 10))\n",
    "sns.heatmap(\n",
    "    corr_matrix,\n",
    "    annot=True,\n",
    "    fmt=\".2f\",\n",
    "    cmap=\"vlag\",\n",
    "    center=0,\n",
    "    square=True,\n",
    "    linewidths=0.5,\n",
    "    cbar_kws={\"shrink\": 0.8},\n",
    ")\n",
    "plt.title(\"Pairwise Pearson Correlation of Electrophysiological Features\")\n",
    "plt.xticks(rotation=45, ha=\"right\")\n",
    "plt.yticks(rotation=0)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*Interpretation:*\n",
    "\n",
    "Positive correlations:\n",
    "* AP width (ms) x Upstroke-to-downstroke ratio: +0.94. Suggests these may capture overlapping aspects of action potential shape.\n",
    "* Afterhyperpolarization (mV) x Upstroke-to-downstroke ratio: +0.67. Larger APs may lead to deeper afterhyperpolarizations, consistent with biophysical expectations\n",
    "* AP amplitude (mV) x Afterhyperpolarization (mV): +0.67. Larger APs may lead to deeper afterhyperpolarizations, possibly reflecting cell type-specific properties or biophysical mechanisms.\n",
    "\n",
    "Negative correlations:\n",
    "* AP width (ms) x Max number of APs: -0.73. Cells firining more APs tend to have narrower APs, possibly indicating a trade-off between firing rate and action potential width.\n",
    "* ISI coefficient of variation x Max number of APs: -0.70. More consistent ISIs are found in cells with higher firing rates, hinting at firing pattern specialization.\n",
    "* Max number of APs x Upstroke-to-downstroke ratio: -0.68. Suggests that cells with more APs tend to have lower upstroke-to-downstroke ratios, possibly indicating differences in firing patterns or cell types.\n",
    "\n",
    "Generally, highly correlated features might be redundant or be functionally related, which will be the main part of the next tasks.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1.4 Investigate Ground Truth Labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Analyze and visualize ground truth distribution\n",
    "# ground truth data:\n",
    "# - rna_type: 77 trancriptomic subtypes\n",
    "# - rna_family: 9 broader transcriptomic families\n",
    "\n",
    "# Barplot of RNA types\n",
    "rna_type_counts = pd.Series(rna_type).value_counts()\n",
    "\n",
    "plt.figure(figsize=(12, 5))\n",
    "sns.barplot(\n",
    "    x=rna_type_counts.index,\n",
    "    y=rna_type_counts.values,\n",
    "    hue=rna_type_counts.index,\n",
    "    palette=\"muted\",\n",
    "    dodge=False,\n",
    "    legend=False,\n",
    ")\n",
    "plt.xticks(rotation=90)\n",
    "plt.title(\"RNA Types by Cell Count\")\n",
    "plt.ylabel(\"Number of Cells\")\n",
    "plt.xlabel(\"RNA Type\")\n",
    "plt.show()\n",
    "\n",
    "# Barplot of RNA families\n",
    "plt.figure(figsize=(8, 4))\n",
    "sns.countplot(\n",
    "    x=rna_family,\n",
    "    hue=rna_family,\n",
    "    order=np.unique(rna_family),\n",
    "    palette=\"Set2\",\n",
    "    legend=False,\n",
    ")\n",
    "plt.title(\"Distribution of RNA Families\")\n",
    "plt.ylabel(\"Number of Cells\")\n",
    "plt.xlabel(\"RNA Family\")\n",
    "plt.xticks(rotation=45)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*Interpretation:*\n",
    "\n",
    "Top: RNA types by cell count\n",
    "* barplot shows the number of cells annotated with each RNA type\n",
    "* only a few rna types are considerably highly represented (e.g.Pvalb Itrap2, L2/3 IT_2/)\n",
    "* many RNA types have fewer than 10 cells, suggesting class imbalance --> could impact clustering and classification methods \n",
    "\n",
    "Bottom: RNA Families by cell count\n",
    "* the plot aggregates rna types into broader families, which are essentially coarser labels\n",
    "* Families like IT, Pvalb and Sst are dominating, whereas others like NP and Sncp are very underrepresented\n",
    "* Compared to the top plot, the family distribution is more balanced, with fewer families having very few cells, yet the class imbalance is still present\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1.5 Metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# print metadata labels\n",
    "print(\"Metadata labels:\")\n",
    "for col in meta.columns:\n",
    "    print(f\"- {col}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Generally, we want to focus on metadata that improves analysis quality. Not all recorded metadata is useful for our purposes, so we inspected the distributions and properties of the metadata. \n",
    "\n",
    "We additionally performed a background research which resulted in the following features:\n",
    "\n",
    "Cre Line (Tasic et al. 2016):\n",
    "* Cre lines indicate genetic targeting strategies, which can enrich for specific neuron types\n",
    "* They help validate whether observed patterns are biologically driven or due to experimental targeting\n",
    "\n",
    "Inferred layer (Loo et al. 2019):\n",
    "* Cortical layers correspond to distinct cell populations and functions; many electrophysiological and transcriptomic differences are layer-specific.\n",
    "* It could serve as a meaningful covariate or stratification variable\n",
    "\n",
    "Sequencing batch (Tung et al. 2017):\n",
    "* Sequencing batch can introduce technical variability (batch effects) unrelated to biology\n",
    "* Identifying batch effects can help to avoid spurious conclusions and ensure generalizability\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Distribution of Cre lines\n",
    "cre_counts = meta[\"Cre\"][exclude_low_quality].value_counts()\n",
    "\n",
    "# Plot\n",
    "plt.figure(figsize=(10, 4))\n",
    "sns.countplot(\n",
    "    y=meta[\"Cre\"][exclude_low_quality],\n",
    "    order=cre_counts.index,\n",
    "    palette=\"muted\",\n",
    "    hue=meta[\"Cre\"][exclude_low_quality],\n",
    "    legend=False,\n",
    ")\n",
    "plt.title(\"Distribution of Cre Lines\")\n",
    "plt.xlabel(\"Cell Count\")\n",
    "plt.ylabel(\"Cre Line\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Distribution of cortical layers\n",
    "layer_counts = meta[\"Inferred layer\"][exclude_low_quality].value_counts()\n",
    "\n",
    "# Plot\n",
    "plt.figure(figsize=(8, 4))\n",
    "sns.countplot(\n",
    "    x=meta[\"Inferred layer\"][exclude_low_quality],\n",
    "    order=layer_counts.index,\n",
    "    palette=\"coolwarm\",\n",
    "    hue=meta[\"Inferred layer\"][exclude_low_quality],\n",
    "    legend=False,\n",
    ")\n",
    "plt.title(\"Distribution of Inferred Cortical Layers\")\n",
    "plt.xlabel(\"Inferred Layer\")\n",
    "plt.ylabel(\"Number of Cells\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Distribution of batches\n",
    "batch_counts = meta[\"Sequencing batch\"][exclude_low_quality].value_counts()\n",
    "\n",
    "# plot\n",
    "plt.figure(figsize=(8, 4))\n",
    "sns.barplot(\n",
    "    x=batch_counts.index.astype(str),\n",
    "    y=batch_counts.values,\n",
    "    palette=\"coolwarm\",\n",
    "    hue=batch_counts.index.astype(str),\n",
    "    legend=False,\n",
    ")\n",
    "plt.title(\"Distribution of Sequencing Batches\")\n",
    "plt.xlabel(\"Sequencing Batch (ordered by cell count)\")\n",
    "plt.ylabel(\"Cell Count\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Metadata Conclusions\n",
    "\n",
    "Cre Lines:\n",
    "* The distribution shows several dominant lines (e.g. SST+, PV+), with many others represented sparsely.\n",
    "* A large number of unique Cre lines suggests some are too underrepresented for individual modeling but may help explain biological variability when grouped or used selectively.\n",
    "\n",
    "Inferred layer:\n",
    "* The distribution is uneven, with most cells in layer 5, followed by 2/3 and 6, and very few in layer 1.\n",
    "* This may impact how well certain layers are represented in modeling, and highlights that layer-specific trends could be explored in later tasks.\n",
    "\n",
    "Sequencing batch:\n",
    "* The barplot revealed an uneven distribution across batches, ranging from very large (e.g. batch 8) to tiny (batch 12).\n",
    "\n",
    "**How they could be used in further analyses?:**\n",
    "\n",
    "Cre Lines:\n",
    "* use after clustering --> compare predicted clusters to Cre line distributions\n",
    "* helps biologically validate clusters, e.g. a cluster dominated by SST+ Cre lines might represents inhibitory neurons (task 4)\n",
    "* can examine how cre lines associate with PCs or gene expression signatures (task 5)\n",
    "\n",
    "Inferred layer:\n",
    "* use inferred layer as a coloring option to see if spatial cell identity is reflected in transcriptomic space (task 3)\n",
    "* could be added as a covariate or group in regression or correlation analyses to see how gen-PC correlations differ across layers (task 5)\n",
    "\n",
    "Sequencing batch:\n",
    "* use batch information to check for batch effects after normalization and transformation \n",
    "* could color 2D visualizations (e.g. PCA/t-SNE) by batch to visually inspect clustering by batch, which is undesired and indicates technical confounding (task 2)\n",
    "* use batch labels as negative control labels: ideally batches should not be separable in a meaningful embedding (task 3)\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Task 2\n",
    "\n",
    "Preparation steps:\n",
    "\n",
    "* we want to use meta data\n",
    "* therefore we need to filter the metadata to only include the cells we are interested in\n",
    "* this is done by excluding the low quality cells which is only already done for rna_type, rna_family and exonCounts\n",
    "* to avoid misalignment, we reset indices so that the indices of the meta data match the indices of the exonCounts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# preparation steps\n",
    "\n",
    "# RNA counts\n",
    "exonCounts_filtered = exonCounts\n",
    "\n",
    "# Ground truth labels\n",
    "rna_type_filtered = rna_type.reset_index(drop=True)\n",
    "rna_family_filtered = rna_family.reset_index(drop=True)\n",
    "\n",
    "# Filter Metadata variables of interest\n",
    "cre_filtered = meta[\"Cre\"][exclude_low_quality].reset_index(drop=True)\n",
    "layer_filtered = meta[\"Inferred layer\"][exclude_low_quality].reset_index(\n",
    "    drop=True\n",
    ")\n",
    "batch_filtered = meta[\"Sequencing batch\"][exclude_low_quality].reset_index(\n",
    "    drop=True\n",
    ")\n",
    "\n",
    "assert (\n",
    "    exonCounts_filtered.shape[0]\n",
    "    == len(rna_type_filtered)\n",
    "    == len(cre_filtered)\n",
    "    == len(layer_filtered)\n",
    "    == len(batch_filtered)\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.1 normalization + transformation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "log-CPM normalization - appropriately tailored for read count data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# compute total amount of reads per cell\n",
    "total_counts_per_cell = exonCounts_filtered.sum(axis=1)\n",
    "\n",
    "# avoid dividing by zero by setting any 0s to 1\n",
    "total_counts_per_cell[total_counts_per_cell == 0] = 1\n",
    "\n",
    "# NORMALIZATION\n",
    "# calculate counts per million\n",
    "cpm = (\n",
    "    exonCounts_filtered.T / total_counts_per_cell\n",
    ").T * 1e6  # transpose to divide along columns (genes)\n",
    "\n",
    "# TRANSFORMATION\n",
    "# log-transform with log(1+x)\n",
    "log_cpm = np.log1p(cpm)\n",
    "# sqrt-transform with sqrt(x)\n",
    "sqrt_cpm = np.sqrt(cpm)\n",
    "\n",
    "# create dataframe\n",
    "cpm_df = pd.DataFrame(cpm, index=rna_type_filtered, columns=genes)\n",
    "log_cpm_df = pd.DataFrame(log_cpm, index=rna_type_filtered, columns=genes)\n",
    "sqrt_cpm_df = pd.DataFrame(sqrt_cpm, index=rna_type_filtered, columns=genes)\n",
    "\n",
    "print(\"Log-CPM shape:\", log_cpm.shape)\n",
    "print(\"Sqrt-CPM shape:\", sqrt_cpm.shape)\n",
    "\n",
    "assert log_cpm_df.shape[0] == len(rna_type_filtered), \"check for alignment\"\n",
    "assert sqrt_cpm_df.shape[0] == len(rna_type_filtered), \"check for alignment\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# show how normalization changes total RNA counts\n",
    "fig, axes = plt.subplots(1, 2, figsize=(10, 4))\n",
    "\n",
    "axes[0].bar(\n",
    "    range(len(total_counts_per_cell)),\n",
    "    total_counts_per_cell,\n",
    "    width=1,\n",
    "    alpha=0.5,\n",
    "    color=\"blue\",\n",
    ")\n",
    "axes[0].set_yscale(\"log\")\n",
    "axes[0].set_xlabel(\"Cell index\")\n",
    "axes[0].set_ylabel(\"Raw RNA counts\")\n",
    "axes[0].set_title(\"Total RNA Counts Per Cell\")\n",
    "\n",
    "total_cpm_counts_per_cell = cpm.sum(axis=1)\n",
    "\n",
    "\n",
    "axes[1].bar(\n",
    "    range(len(total_cpm_counts_per_cell)),\n",
    "    total_cpm_counts_per_cell,\n",
    "    width=1,\n",
    "    alpha=0.5,\n",
    "    color=\"blue\",\n",
    ")\n",
    "axes[1].set_yscale(\"log\")\n",
    "axes[1].set_xlabel(\"Cell index\")\n",
    "axes[1].set_ylabel(\"Normalized RNA counts\")\n",
    "axes[1].set_title(\"Total Normalized RNA Counts Per Cell\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "=> the raw counts are larger numbers whereas the log-CPM shows a compressed scale"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# show change after log transform\n",
    "\n",
    "fig, axes = plt.subplots(1, 3, figsize=(10, 4))\n",
    "\n",
    "axes[0].scatter(\n",
    "    mean_expression_across_cells,\n",
    "    variance_expression_across_cells,\n",
    "    s=5,\n",
    "    alpha=0.5,\n",
    ")\n",
    "axes[0].set_xscale(\"log\")\n",
    "axes[0].set_yscale(\"log\")\n",
    "axes[0].set_xlabel(\"Mean Expression Across Cells (log scale)\")\n",
    "axes[0].set_ylabel(\"Variance Across Cells (log scale)\")\n",
    "axes[0].set_title(\"Mean vs Variance of Gene Expression\")\n",
    "\n",
    "\n",
    "mean_log_expression = log_cpm.mean(axis=0)\n",
    "var_log_expression = log_cpm.var(axis=0)\n",
    "\n",
    "\n",
    "axes[1].scatter(\n",
    "    mean_log_expression,\n",
    "    var_log_expression,\n",
    "    s=5,\n",
    "    alpha=0.5,\n",
    ")\n",
    "axes[1].set_xscale(\"log\")\n",
    "axes[1].set_yscale(\"log\")\n",
    "axes[1].set_xlabel(\"Mean Expression Across Cells (log scale)\")\n",
    "axes[1].set_ylabel(\"Variance Across Cells (log scale)\")\n",
    "axes[1].set_title(\"Mean vs Variance of Gene Expression\\nAfter Log-Transform\")\n",
    "\n",
    "# show change after sqrt transform\n",
    "mean_sqrt_expression = sqrt_cpm.mean(axis=0)\n",
    "var_sqrt_expression = sqrt_cpm.var(axis=0)\n",
    "\n",
    "\n",
    "axes[2].scatter(\n",
    "    mean_sqrt_expression,\n",
    "    var_sqrt_expression,\n",
    "    s=5,\n",
    "    alpha=0.5,\n",
    ")\n",
    "axes[2].set_xscale(\"log\")\n",
    "axes[2].set_yscale(\"log\")\n",
    "axes[2].set_xlabel(\"Mean Expression Across Cells (log scale)\")\n",
    "axes[2].set_ylabel(\"Variance Across Cells (log scale)\")\n",
    "axes[2].set_title(\"Mean vs Variance of Gene Expression\\nAfter Sqrt-Transform\")\n",
    "\n",
    "# add identity lines\n",
    "x_vals = np.linspace(1e-3, 1e4, 100)  # log scale range\n",
    "axes[0].plot(\n",
    "    x_vals, x_vals, linestyle=\"--\", color=\"red\", label=\"variance = mean\"\n",
    ")\n",
    "x_vals = np.linspace(1e-4, 1e1, 100)  # log scale range\n",
    "axes[1].plot(\n",
    "    x_vals, x_vals, linestyle=\"--\", color=\"red\", label=\"variance = mean\"\n",
    ")\n",
    "x_vals = np.linspace(1e-4, 1e2, 100)  # log scale range\n",
    "axes[2].plot(\n",
    "    x_vals, x_vals, linestyle=\"--\", color=\"red\", label=\"variance = mean\"\n",
    ")\n",
    "\n",
    "axes[0].legend()\n",
    "axes[1].legend()\n",
    "axes[2].legend()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*Interpretation:*\n",
    "\n",
    "Left: Raw CPM (before transformation)\n",
    "* variance increases rapidly with mean, indicating overdispersion\n",
    "* many genes show much higher variance than mean, deviating strongly from the Poisson assumption variance = mean. But this is expected for raw count data, as it is not supposed to follow a Poisson distribution.\n",
    "* Strong heteroscedasticity (variance depends on the magnitude of the mean). Could hint that PCA might not be the best choice for this data, as PCA assumes homoscedasticity (constant variance across all means).\n",
    "\n",
    "Middle: after log transformation\n",
    "* cloud of points is more compact, with less spread in variance for high means\n",
    "* points are now more evenly distributed around the identity line, indicating that the log transformation has stabilized variance and made the data more homoscedastic (e.g. compress extreme values and reduce skew)\n",
    "* log transformed data looks now more suitable for PCA\n",
    "\n",
    "Right: after sqrt transformation\n",
    "* despite not changing the overall shape much, sqrt transform has also stabilized variance and made the data more homoscedastic\n",
    "* sqrt transformed data also looks more suitable for PCA"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.2 select genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# choose highly variable genes (hvgs) from normalized data\n",
    "# we DO NOT use log-transformed data here, since that would flatten the mean-variance relationship\n",
    "# (and we want to find the genes with high absolute variance)\n",
    "\n",
    "# mask out genes with zero mean\n",
    "mask_raw = mean_expression_across_cells > 0.1\n",
    "\n",
    "dispersion = variance_expression_across_cells / (\n",
    "    mean_expression_across_cells + 1e-8\n",
    ")\n",
    "\n",
    "# remove genes with very low mean (avoid divide by zero)\n",
    "filtered_dispersion = dispersion[mask_raw]\n",
    "filtered_mean = mean_expression_across_cells[mask_raw]\n",
    "\n",
    "\n",
    "# 2000 HVGs is the most common default, but even 1000 lead to a very high necessary number of principle components in PCA later on.\n",
    "# Fewer than 500 HVGs risks missing subtle biological signals, so we chose 500 as the lowest \"possible\" number.\n",
    "\n",
    "# We also looked at 1000 HVGs to compare it later on with the 500 HVGs.\n",
    "\n",
    "top_500_indices = np.argsort(filtered_dispersion)[-500:]\n",
    "top_1000_indices = np.argsort(filtered_dispersion)[-1000:]\n",
    "\n",
    "# get gene names\n",
    "hvg_genes_500 = np.array(genes)[mask_raw][top_500_indices]\n",
    "hvg_genes_1000 = np.array(genes)[mask_raw][top_1000_indices]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot mean-variance with hvgs highlighted\n",
    "\n",
    "# mask of genes that are in hvg list\n",
    "is_hvg_500 = np.isin(genes, hvg_genes_500)\n",
    "is_hvg_1000 = np.isin(genes, hvg_genes_1000)\n",
    "\n",
    "# fig, ax = plt.subplots(figsize=(6, 5))\n",
    "fig, axes = plt.subplots(1, 2, figsize=(12, 4))\n",
    "\n",
    "# for 500 hvgs\n",
    "# all genes\n",
    "axes[0].scatter(\n",
    "    mean_expression_across_cells[mask_raw],\n",
    "    variance_expression_across_cells[mask_raw],\n",
    "    s=5,\n",
    "    alpha=0.4,\n",
    "    label=\"all genes\",\n",
    "    color=\"lightblue\",\n",
    ")\n",
    "\n",
    "# overlay HVGs\n",
    "axes[0].scatter(\n",
    "    mean_expression_across_cells[is_hvg_500],\n",
    "    variance_expression_across_cells[is_hvg_500],\n",
    "    s=10,\n",
    "    alpha=0.7,\n",
    "    label=f\"Top 500 HVGs by dispersion\",\n",
    "    color=\"darkblue\",\n",
    ")\n",
    "\n",
    "# for 1000 hvgs\n",
    "# all genes\n",
    "axes[1].scatter(\n",
    "    mean_expression_across_cells[mask_raw],\n",
    "    variance_expression_across_cells[mask_raw],\n",
    "    s=5,\n",
    "    alpha=0.4,\n",
    "    label=\"all genes\",\n",
    "    color=\"lightblue\",\n",
    ")\n",
    "\n",
    "# overlay HVGs\n",
    "axes[1].scatter(\n",
    "    mean_expression_across_cells[is_hvg_1000],\n",
    "    variance_expression_across_cells[is_hvg_1000],\n",
    "    s=10,\n",
    "    alpha=0.7,\n",
    "    label=f\"Top 1000 HVGs by dispersion\",\n",
    "    color=\"darkblue\",\n",
    ")\n",
    "\n",
    "for ax in axes:\n",
    "    ax.set_xscale(\"log\")\n",
    "    ax.set_yscale(\"log\")\n",
    "    ax.set_xlabel(\"Mean Expression Across Cells\")\n",
    "    ax.set_ylabel(\"Variance Across Cells\")\n",
    "    ax.set_title(\"HVGs Highlighted in Mean–Variance Plot\")\n",
    "    ax.legend()\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.3 PCA\n",
    "\n",
    "note: use normalized + log-transformed data here!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# find indices of HVGs\n",
    "hvg_indices_500 = np.where(np.isin(genes, hvg_genes_500))[0]\n",
    "hvg_indices_1000 = np.where(np.isin(genes, hvg_genes_1000))[0]\n",
    "\n",
    "# subset the expression matrix: shape = (cells, HVGs)\n",
    "hvg_expression_500 = exonCounts[:, hvg_indices_500]\n",
    "hvg_expression_1000 = exonCounts[:, hvg_indices_1000]\n",
    "\n",
    "# use normalized + transformed data from before\n",
    "log_hvg_df_500 = log_cpm_df.iloc[:, hvg_indices_500].copy()\n",
    "sqrt_hvg_df_500 = sqrt_cpm_df.iloc[:, hvg_indices_500].copy()\n",
    "log_hvg_df_1000 = log_cpm_df.iloc[:, hvg_indices_1000].copy()\n",
    "sqrt_hvg_df_1000 = sqrt_cpm_df.iloc[:, hvg_indices_1000].copy()\n",
    "\n",
    "# non-transformed data\n",
    "raw_hvg_df_500 = cpm_df.iloc[:, hvg_indices_500].copy()\n",
    "raw_hvg_df_1000 = cpm_df.iloc[:, hvg_indices_1000].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# fit PCA + play with the number of components\n",
    "# 400 components was shown to be the lowest number that still explains at least 60% of variance in the data\n",
    "pca_log_500 = PCA(\n",
    "    n_components=400  \n",
    ")\n",
    "pca_log_1000 = PCA(\n",
    "    n_components=400 \n",
    ")\n",
    "X_pca_log_500 = pca_log_500.fit_transform(log_hvg_df_500)\n",
    "X_pca_log_1000 = pca_log_1000.fit_transform(log_hvg_df_1000)\n",
    "\n",
    "pca_sqrt_500 = PCA(\n",
    "    n_components=400  \n",
    ")\n",
    "pca_sqrt_1000 = PCA(\n",
    "    n_components=400  \n",
    ")\n",
    "X_pca_sqrt_500 = pca_sqrt_500.fit_transform(sqrt_hvg_df_500)\n",
    "X_pca_sqrt_1000 = pca_sqrt_1000.fit_transform(sqrt_hvg_df_1000)\n",
    "\n",
    "# non-transformed data\n",
    "pca_raw_500 = PCA(\n",
    "    n_components=400  \n",
    ")\n",
    "pca_raw_1000 = PCA(\n",
    "    n_components=400  \n",
    ")\n",
    "X_pca_raw_500 = pca_raw_500.fit_transform(raw_hvg_df_500)\n",
    "X_pca_raw_1000 = pca_raw_1000.fit_transform(raw_hvg_df_1000)\n",
    "\n",
    "# Check if the number of cells matches the number of RNA types\n",
    "assert (\n",
    "    len(rna_type_filtered) == X_pca_log_500.shape[0]\n",
    "), \"Mismatch between RNA types and PCA output\"\n",
    "\n",
    "print(\"PCA result shape:\", X_pca_log_500.shape)  # (num_cells, n_components)\n",
    "print(\"PCA result shape:\", X_pca_sqrt_1000.shape)  # (num_cells, n_components)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot explained variance to compare\n",
    "explained_log_500 = pca_log_500.explained_variance_ratio_\n",
    "explained_sqrt_500 = pca_sqrt_500.explained_variance_ratio_\n",
    "explained_log_1000 = pca_log_1000.explained_variance_ratio_\n",
    "explained_sqrt_1000 = pca_sqrt_1000.explained_variance_ratio_\n",
    "\n",
    "explained_raw_500 = pca_raw_500.explained_variance_ratio_\n",
    "explained_raw_1000 = pca_raw_1000.explained_variance_ratio_\n",
    "\n",
    "fig, axes = plt.subplots(1, 2, figsize=(12, 4))\n",
    "\n",
    "axes[0].plot(\n",
    "    np.cumsum(explained_log_1000), color=\"blue\", label=\"log (1000 HVGs)\"\n",
    ")\n",
    "axes[0].plot(\n",
    "    np.cumsum(explained_sqrt_1000), color=\"orange\", label=\"sqrt (1000 HVGs)\"\n",
    ")\n",
    "axes[0].plot(\n",
    "    np.cumsum(explained_raw_1000), color=\"green\", label=\"raw (1000 HVGs)\"\n",
    ")\n",
    "axes[1].plot(\n",
    "    np.cumsum(explained_log_500), color=\"blue\", label=\"log (500 HVGs)\"\n",
    ")\n",
    "axes[1].plot(\n",
    "    np.cumsum(explained_sqrt_500), color=\"orange\", label=\"sqrt (500 HVGs)\"\n",
    ")\n",
    "axes[1].plot(\n",
    "    np.cumsum(explained_raw_500), color=\"green\", label=\"raw (500 HVGs)\"\n",
    ")\n",
    "\n",
    "axes[0].axhline(0.9, color=\"red\", linestyle=\"--\", label=\"90% threshold\")\n",
    "axes[1].axhline(0.9, color=\"red\", linestyle=\"--\", label=\"90% threshold\")\n",
    "\n",
    "for ax in axes:\n",
    "    ax.set_xlabel(\"Number of Components\")\n",
    "    ax.set_ylabel(\"Cumulative Explained Variance\")\n",
    "    ax.set_title(\"Explained Variance by PCA Components\")\n",
    "    ax.legend()\n",
    "    ax.grid(True)\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Looking at raw data first, the curve is very steep and the first few components already explain most of the variance. However, this is most likely not biologically meaningful variance, since raw CPM values have dominant technical noise like sequencing depth or cell size. For this reason, we chose not to use the raw data in the following analyses.\n",
    "\n",
    "For 1000 HVGs and log-transform, 390 components would be needed to capture ~90% of the variance. Sqrt-transform performs only a little better, around 275 components would be needed here. Even for 500 HVGS, way above 100 PCs are needed to explain 90% of the variance. This means that the variance is spread out across many dimensions and most individual PCs explain only a small amount of variance. Even though more PCs can capture more information, most of that may not be biologically meaningful, so choosing around fewer PCs still makes more sense.\n",
    "\n",
    "While sqrt-transformed data seems a little bit better at capturing variance, we cannot be sure whether that variance is meaningful or noise. Since log-transform is more commonly used for bio data and better at reducing noise, we will mostle use that in the following. For 500 HVGs, ... PCs can account for ~80% of the variance, which seems fair given that it’s better to retain fewer PCs that reflect meaningful signal than to overfit with noise. Still, sqrt might come in handy for e.g. UMAP visualizations."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter RNA types and families for visualization\n",
    "labels_rna_family_filtered = LabelEncoder().fit_transform(rna_family_filtered)\n",
    "\n",
    "fig, axes = plt.subplots(1, 2, figsize=(12, 5))\n",
    "\n",
    "# plot the first 2 components for rna type and family\n",
    "# PCA colored by RNA Type\n",
    "# Get color values for the filtered RNA types\n",
    "rna_type_colors_filtered = np.vectorize(dict_rna_type_colors.get)(\n",
    "    rna_type_filtered\n",
    ")\n",
    "\n",
    "scatter1 = axes[0].scatter(\n",
    "    X_pca_log_500[:, 0],\n",
    "    X_pca_log_500[:, 1],\n",
    "    c=rna_type_colors_filtered,\n",
    "    s=5,\n",
    "    alpha=0.6,\n",
    ")\n",
    "axes[0].set_title(\"Colored by RNA Type\")\n",
    "axes[0].set_xlabel(\"PC1\")\n",
    "axes[0].set_ylabel(\"PC2\")\n",
    "axes[0].grid(True)\n",
    "\n",
    "# PCA colored by RNA Family\n",
    "scatter2 = axes[1].scatter(\n",
    "    X_pca_log_500[:, 0],\n",
    "    X_pca_log_500[:, 1],\n",
    "    c=labels_rna_family_filtered,\n",
    "    cmap=\"tab10\",\n",
    "    s=5,\n",
    "    alpha=0.6,\n",
    ")\n",
    "axes[1].set_title(\"Colored by RNA Family\")\n",
    "axes[1].set_xlabel(\"PC1\")\n",
    "axes[1].set_ylabel(\"PC2\")\n",
    "axes[1].grid(True)\n",
    "plt.suptitle(\"PCA of Highly Variable Genes (HVGs)\")\n",
    "\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set parameters\n",
    "hvg_counts = [500, 1000, 2000]\n",
    "\n",
    "\n",
    "# Set up figure\n",
    "fig, axes = plt.subplots(len(hvg_counts), 2, figsize=(10, 9))\n",
    "fig.suptitle(\n",
    "    \"Effect of HVG Count on PCA – RNA Type vs RNA Family\", fontsize=16\n",
    ")\n",
    "\n",
    "# Loop through combinations\n",
    "for i, n_hvg in enumerate(hvg_counts):\n",
    "    # HVG selection\n",
    "    dispersion = variance_expression_across_cells / (\n",
    "        mean_expression_across_cells + 1e-8\n",
    "    )\n",
    "    mask = mean_expression_across_cells > 0.1\n",
    "\n",
    "    filtered_dispersion = dispersion[mask]\n",
    "    top_indices = np.argsort(filtered_dispersion)[-n_hvg:]\n",
    "    hvg_genes = np.array(genes)[mask][top_indices]\n",
    "\n",
    "    # Subset & scale\n",
    "    log_hvg_df = log_cpm_df[hvg_genes]\n",
    "    data = StandardScaler().fit_transform(log_hvg_df.values)\n",
    "\n",
    "    # PCA to 2D\n",
    "    pca = PCA(n_components=2)\n",
    "    X_pca = pca.fit_transform(data)\n",
    "\n",
    "    # Plot rna_type\n",
    "    axes[i, 0].scatter(\n",
    "        X_pca[:, 0], X_pca[:, 1], c=rna_type_colors, s=10, alpha=0.6\n",
    "    )\n",
    "    axes[i, 0].set_title(f\"{n_hvg} HVGs\\nColored by RNA Type\")\n",
    "    axes[i, 0].set_xlabel(\"PC1\")\n",
    "    axes[i, 0].set_ylabel(\"PC2\")\n",
    "\n",
    "    # Plot rna_family\n",
    "    axes[i, 1].scatter(\n",
    "        X_pca[:, 0],\n",
    "        X_pca[:, 1],\n",
    "        c=labels_rna_family_filtered,\n",
    "        cmap=\"tab10\",\n",
    "        s=10,\n",
    "        alpha=0.6,\n",
    "    )\n",
    "    axes[i, 1].set_title(f\"{n_hvg} HVGs\\nColored by RNA Family\")\n",
    "    axes[i, 1].set_xlabel(\"PC1\")\n",
    "    axes[i, 1].set_ylabel(\"PC2\")\n",
    "\n",
    "plt.tight_layout(rect=[0, 0, 1, 0.95])\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For 500 HVGs, the plots seem to show the clearest separation between two clusters. This makes sense, since for the least amount of HVGs, the least amount of information is left to be broken down. For more HVGs, 2 PCs can capture even less variance, so the plots look a little more 'chaotic'."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Task 3"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3.1 Visualization"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.1.1 t-SNE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# prepare data\n",
    "log_hvg_df = log_cpm_df[hvg_genes_500]\n",
    "sqrt_hvg_df = sqrt_cpm_df[hvg_genes_500]\n",
    "X_hvg = StandardScaler().fit_transform(log_hvg_df.values)\n",
    "X_hvg_sqrt = StandardScaler().fit_transform(sqrt_hvg_df.values)\n",
    "\n",
    "# sanity checks\n",
    "assert X_hvg.shape[0] == len(rna_type_filtered)\n",
    "assert X_hvg.shape[0] == len(rna_family_filtered)\n",
    "\n",
    "# filter log-normalized gene expression to the same 1224 cells\n",
    "X_hvg_filtered = X_hvg[keepcells, :]  ### hier dann gar nicht verwendet?\n",
    "\n",
    "# define dimensionality reduction methods and parameters\n",
    "methods = {\n",
    "    \"t-SNE (perp=5)\": {\n",
    "        \"func\": lambda X: TSNE(\n",
    "            n_components=2, perplexity=5, max_iter=1000, random_state=42\n",
    "        ).fit_transform(X)\n",
    "    },\n",
    "    \"t-SNE (perp=30)\": {\n",
    "        \"func\": lambda X: TSNE(\n",
    "            n_components=2, perplexity=30, max_iter=1000, random_state=42\n",
    "        ).fit_transform(X)\n",
    "    },\n",
    "    \"t-SNE (perp=100)\": {\n",
    "        \"func\": lambda X: TSNE(\n",
    "            n_components=2, perplexity=100, max_iter=1000, random_state=42\n",
    "        ).fit_transform(X)\n",
    "    },\n",
    "}\n",
    "\n",
    "# create subplots: 2 rows (RNA type + RNA family) x 3 columns (t-SNE variants)\n",
    "fig, axes = plt.subplots(2, len(methods), figsize=(16, 9))\n",
    "fig.suptitle(\n",
    "    \"t-SNE Projections Colored by RNA Type and RNA Family\", fontsize=16\n",
    ")\n",
    "\n",
    "for col_idx, (title, conf) in enumerate(methods.items()):\n",
    "    X_embedded = conf[\"func\"](X_hvg)\n",
    "\n",
    "    # row 0: RNA type\n",
    "    sns.scatterplot(\n",
    "        x=X_embedded[:, 0],\n",
    "        y=X_embedded[:, 1],\n",
    "        hue=rna_type_filtered,\n",
    "        palette=dict_rna_type_colors,\n",
    "        s=15,\n",
    "        alpha=0.7,\n",
    "        legend=False,\n",
    "        ax=axes[0, col_idx],\n",
    "    )\n",
    "    axes[0, col_idx].set_title(f\"{title} – RNA Type\")\n",
    "    axes[0, col_idx].set_xlabel(\"Dim 1\")\n",
    "    axes[0, col_idx].set_ylabel(\"Dim 2\")\n",
    "\n",
    "    # row 1: RNA family\n",
    "    sns.scatterplot(\n",
    "        x=X_embedded[:, 0],\n",
    "        y=X_embedded[:, 1],\n",
    "        hue=rna_family_filtered,\n",
    "        palette=\"tab10\",\n",
    "        s=15,\n",
    "        alpha=0.7,\n",
    "        legend=False,\n",
    "        ax=axes[1, col_idx],\n",
    "    )\n",
    "    axes[1, col_idx].set_title(f\"{title} – RNA Family\")\n",
    "    axes[1, col_idx].set_xlabel(\"Dim 1\")\n",
    "    axes[1, col_idx].set_ylabel(\"Dim 2\")\n",
    "\n",
    "plt.tight_layout(rect=[0, 0, 1, 0.95])\n",
    "plt.show()\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The comparison shows how t-SNE's perplexity parameter affects local versus global structure:\n",
    "* perplexity=5: clusters are small and tightly packed => emphasis on fine local relationships\n",
    "* perplexity=30: balance between local detail & global structure => seems to provide the best overview of the data\n",
    "* perplexity=100: the plot appears more globally smoothed and there is less distinct clustering => highlighting broader structure at the expense of local detail"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.1.2 UMAP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import umap.umap_ as umap\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "# define dimensionality reduction methods and parameters\n",
    "methods = {\n",
    "    \"UMAP (n=15)\": lambda X: umap.UMAP(\n",
    "        n_neighbors=15, min_dist=0.1, random_state=42\n",
    "    ).fit_transform(X),\n",
    "    \"UMAP (n=30)\": lambda X: umap.UMAP(\n",
    "        n_neighbors=30, min_dist=0.3, random_state=42\n",
    "    ).fit_transform(X),\n",
    "    \"UMAP (n=50)\": lambda X: umap.UMAP(\n",
    "        n_neighbors=50, min_dist=0.3, random_state=42\n",
    "    ).fit_transform(X),\n",
    "}\n",
    "\n",
    "# create 2 rows (log + sqrt), n columns = number of methods\n",
    "n_methods = len(methods)\n",
    "fig, axes = plt.subplots(2, n_methods, figsize=(5 * n_methods, 10))\n",
    "axes = axes.reshape(2, n_methods)\n",
    "\n",
    "# plot for log-transformed data\n",
    "for col, (title, func) in enumerate(methods.items()):\n",
    "    X_embedded_log = func(X_hvg)\n",
    "    sns.scatterplot(\n",
    "        x=X_embedded_log[:, 0],\n",
    "        y=X_embedded_log[:, 1],\n",
    "        hue=rna_family,\n",
    "        palette=\"tab10\",\n",
    "        s=30,\n",
    "        alpha=0.7,\n",
    "        legend=False,\n",
    "        ax=axes[0, col],\n",
    "    )\n",
    "    axes[0, col].set_title(f\"{title} — log(CPM)\")\n",
    "    axes[0, col].set_xlabel(\"Dim 1\")\n",
    "    axes[0, col].set_ylabel(\"Dim 2\")\n",
    "\n",
    "# plot for sqrt-transformed data\n",
    "for col, (title, func) in enumerate(methods.items()):\n",
    "    X_embedded_sqrt = func(X_hvg_sqrt)\n",
    "    sns.scatterplot(\n",
    "        x=X_embedded_sqrt[:, 0],\n",
    "        y=X_embedded_sqrt[:, 1],\n",
    "        hue=rna_family,\n",
    "        palette=\"tab10\",\n",
    "        s=30,\n",
    "        alpha=0.7,\n",
    "        legend=False,\n",
    "        ax=axes[1, col],\n",
    "    )\n",
    "    axes[1, col].set_title(f\"{title} — sqrt(CPM)\")\n",
    "    axes[1, col].set_xlabel(\"Dim 1\")\n",
    "    axes[1, col].set_ylabel(\"Dim 2\")\n",
    "\n",
    "plt.suptitle(\n",
    "    \"UMAP Embeddings (log vs sqrt CPM) by Different Neighborhood Sizes\",\n",
    "    fontsize=16,\n",
    "    y=1.03,\n",
    ")\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The comparison shows how the 'n_neighbors' parameter and the type of data transformation affect the balance between local and global structure:\n",
    "\n",
    "**type of data transformation:**\n",
    "\n",
    "Log-transformed and sqrt-transformed data actually look very similar, so it seems like we gain no advantage from the sqrt-transformed data for UMAP, as speculated above. For this reason and for simplicity, we will fully focus on the log-transformed data from now on.\n",
    "\n",
    "**number of neighbors:**\n",
    "\n",
    "n=15: the plot reveals finer substructures and potential cell states\n",
    "\n",
    "n=30 / n=50:\n",
    "* clusters are more connected or blended.\n",
    "* increasing `n_neighbors` favors global structure, but makes us lose finer distinctions\n",
    "* useful to highlight broad cell types instead of subtypes\n",
    "\n",
    "=> smaller 'n_neighbors' values are better for identifying subpopulations, while larger values show more global cell type structure\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.1.3 visual comparison of PCA, t-SNE, UMAP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# define dimensionality reduction methods and parameters\n",
    "methods = {\n",
    "    \"PCA\": {\"func\": lambda X: PCA(n_components=2).fit_transform(X)},\n",
    "    \"t-SNE (perp=30)\": {\n",
    "        \"func\": lambda X: TSNE(\n",
    "            n_components=2, perplexity=30, max_iter=1000, random_state=42\n",
    "        ).fit_transform(X)\n",
    "    },\n",
    "    \"UMAP (n=15)\": {\n",
    "        \"func\": lambda X: umap.UMAP(\n",
    "            n_neighbors=15, min_dist=0.1, random_state=42\n",
    "        ).fit_transform(X)\n",
    "    },\n",
    "}\n",
    "\n",
    "# create subplots\n",
    "n_methods = len(methods)\n",
    "fig, axes = plt.subplots(1, n_methods, figsize=(5 * n_methods, 5))\n",
    "\n",
    "# loop through and plot\n",
    "for ax, (title, conf) in zip(axes, methods.items()):\n",
    "    X_embedded = conf[\"func\"](X_hvg)\n",
    "\n",
    "    sns.scatterplot(\n",
    "        x=X_embedded[:, 0],\n",
    "        y=X_embedded[:, 1],\n",
    "        palette=\"tab10\",\n",
    "        s=30,\n",
    "        alpha=0.7,\n",
    "        legend=False,\n",
    "        ax=ax,\n",
    "    )\n",
    "\n",
    "    ax.set_title(title)\n",
    "    ax.set_xlabel(\"Dim 1\")\n",
    "    ax.set_ylabel(\"Dim 2\")\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Interpretation across visualization types: \n",
    "\n",
    "PCA\n",
    "\n",
    "* no clear clusters.\n",
    "* seems to capture global variance, but fails to separate subtle biological subpopulations\n",
    "=> rather use for noise reduction / preprocessing\n",
    "\n",
    "t-SNE (`n_neighbors`=30)\n",
    "\n",
    "* multiple well-separated clusters.\n",
    "* seems to reveal local structure, separating even small subpopulations\n",
    "* downside: t-SNE is not ideal for preserving global relationships\n",
    "\n",
    "UMAP (`n_neighbors`=15)\n",
    "\n",
    "* distinct clusters, some continuity\n",
    "* seems to balances local and global structure, showing gradients and groupings\n",
    "\n",
    "\n",
    "=> further proceedings:\n",
    "\n",
    "* Use PCA as input to UMAP/t-SNE\n",
    "* Use UMAP or t-SNE for cluster discovery or cell type visualization.\n",
    "\n",
    "\n",
    "TODO:we can try coloring by RNA type, cluster ID, or marker expression to aid interpretation.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3.2 Comparison using Quantitative Metrics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def evaluate_knn_projection(\n",
    "    cpm_df,\n",
    "    hvg_genes,\n",
    "    labels,\n",
    "    n_neighbors=10,\n",
    "    test_size=0.3,\n",
    "    random_state=42,\n",
    "    pca_components_for_tsne_umap=50,\n",
    "):\n",
    "    \"\"\"\n",
    "    Compare dimensionality reduction methods using kNN classification metrics.\n",
    "\n",
    "    Parameters:\n",
    "        cpm_df (DataFrame): CPM-normalized+transformed gene expression (cells × genes)\n",
    "        hvg_genes (list): list of highly variable gene names\n",
    "        labels (array-like): class labels for each cell (e.g., RNA type)\n",
    "        n_neighbors (int): number of neighbors for kNN\n",
    "        test_size (float): train/test split size\n",
    "        random_state (int): reproducibility seed\n",
    "        pca_components_for_tsne_umap (int): number of PCs to feed into t-SNE/UMAP\n",
    "\n",
    "    Returns:\n",
    "        DataFrame with accuracy and recall for each method.\n",
    "    \"\"\"\n",
    "\n",
    "    # standardize input data\n",
    "    X_hvg = StandardScaler().fit_transform(\n",
    "        cpm_df[hvg_genes].values\n",
    "    )\n",
    "    y = np.array(labels)\n",
    "\n",
    "    # precompute PCA-reduced input for non-linear methods\n",
    "    X_pca_for_embedding = PCA(\n",
    "        n_components=pca_components_for_tsne_umap\n",
    "    ).fit_transform(X_hvg)\n",
    "\n",
    "    methods = {\n",
    "        \"High-dimensional (raw)\": lambda X: X,\n",
    "        \"PCA (2D)\": lambda X: PCA(n_components=2).fit_transform(X),\n",
    "        \"PCA (10D)\": lambda X: PCA(n_components=10).fit_transform(X),\n",
    "        \"t-SNE (2D)\": lambda X: TSNE(\n",
    "            n_components=2, perplexity=30, random_state=random_state\n",
    "        ).fit_transform(X),\n",
    "        \"t-SNE (2D from PCA)\": lambda _: TSNE(\n",
    "            n_components=2, perplexity=30, random_state=random_state\n",
    "        ).fit_transform(X_pca_for_embedding),\n",
    "        \"t-SNE (10D)\": lambda X: TSNE(\n",
    "            n_components=10,\n",
    "            perplexity=30,\n",
    "            method=\"exact\",\n",
    "            random_state=random_state,\n",
    "        ).fit_transform(X),\n",
    "        \"t-SNE (10D from PCA)\": lambda _: TSNE(\n",
    "            n_components=10,\n",
    "            perplexity=30,\n",
    "            method=\"exact\",\n",
    "            random_state=random_state,\n",
    "        ).fit_transform(X_pca_for_embedding),\n",
    "        \"UMAP (2D)\": lambda X: umap.UMAP(\n",
    "            n_neighbors=15, min_dist=0.1, random_state=random_state\n",
    "        ).fit_transform(X),\n",
    "        \"UMAP (2D from PCA)\": lambda _: umap.UMAP(\n",
    "            n_neighbors=15, min_dist=0.1, random_state=random_state\n",
    "        ).fit_transform(X_pca_for_embedding),\n",
    "        \"UMAP (10D from PCA)\": lambda _: umap.UMAP(\n",
    "            n_neighbors=15,\n",
    "            min_dist=0.1,\n",
    "            n_components=10,\n",
    "            random_state=random_state,\n",
    "        ).fit_transform(X_pca_for_embedding),\n",
    "    }\n",
    "\n",
    "    results = []\n",
    "    for name, func in methods.items():\n",
    "        try:\n",
    "            X_proj = func(X_hvg)\n",
    "            X_train, X_test, y_train, y_test = train_test_split(\n",
    "                X_proj,\n",
    "                y,\n",
    "                test_size=test_size,\n",
    "                random_state=random_state,\n",
    "            )\n",
    "            print(f\"{name}: X_proj shape = {X_proj.shape}\")\n",
    "\n",
    "            clf = KNeighborsClassifier(n_neighbors=n_neighbors)\n",
    "            clf.fit(X_train, y_train)\n",
    "            acc = clf.score(X_test, y_test)\n",
    "            y_pred = clf.predict(X_test)\n",
    "            avg_recall = recall_score(\n",
    "                y_test, y_pred, average=\"macro\", zero_division=0\n",
    "            )\n",
    "\n",
    "            # compute silhouette score and AMI as in Lause, Berens, Kobak (2024):\n",
    "\n",
    "            # silhouette score\n",
    "            sil = silhouette_score(X_proj, y)\n",
    "            # unsupervised clustering for AMI\n",
    "            n_clusters = len(np.unique(y))\n",
    "            cluster_labels = KMeans(\n",
    "                n_clusters=n_clusters, random_state=random_state\n",
    "            ).fit_predict(X_proj)\n",
    "            ami = adjusted_mutual_info_score(y, cluster_labels)\n",
    "\n",
    "            results.append(\n",
    "                {\n",
    "                    \"Method\": name,\n",
    "                    \"kNN Accuracy\": acc,\n",
    "                    \"kNN Recall (avg)\": avg_recall,\n",
    "                    \"Silhouette Score\": sil,\n",
    "                    \"AMI\": ami,\n",
    "                }\n",
    "            )\n",
    "        except Exception as e:\n",
    "            results.append(\n",
    "                {\n",
    "                    \"Method\": name,\n",
    "                    \"kNN Accuracy\": None,\n",
    "                    \"kNN Recall (avg)\": None,\n",
    "                    \"Silhouette Score\": None,\n",
    "                    \"AMI\": None,\n",
    "                    \"Error\": str(e),\n",
    "                }\n",
    "            )\n",
    "\n",
    "    return pd.DataFrame(results)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"cpm_df shape:\", log_cpm_df.shape)\n",
    "print(\"Sample of columns:\", log_cpm_df.columns[:5])\n",
    "print(\"Sample of hvg_genes:\", hvg_genes[:5])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "results_df = evaluate_knn_projection(\n",
    "    cpm_df=log_cpm_df,\n",
    "    hvg_genes=hvg_genes,\n",
    "    labels=rna_type,\n",
    "    n_neighbors=10,\n",
    "    pca_components_for_tsne_umap=50,\n",
    ")\n",
    "print(results_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "GANZ AM ENDE WERTE NOCHMAL ANPASSEN!\n",
    "\n",
    "Best Overall (based on all metrics): t-SNE (10D from PCA):\n",
    "* kNN Accuracy: 0.46\n",
    "* Recall: 0.26\n",
    "* AMI: 0.44 (highest)\n",
    "* Silhouette: ~-0.21\n",
    "\n",
    "kNN-performance of the models:\n",
    "* t-SNE (10D) is our best performer in terms of accuracy\n",
    "* t-SNE (10D from PCA) has the highest recall\n",
    "* PCA (2D) performs poorly — linear reduction to 2D doesn't seem to work well for this data\n",
    "* UMAP is competitive but doesn't benefit much from going to 10D\n",
    "* Raw high-dimensional data still performs surprisingly well — showing that your data has strong native structure.\n",
    "* a kNN accuracy of around 40% is not high but seems acceptable given the biological data\n",
    "\n",
    "Overall low silhouette scores:\n",
    "All low-dimensional methods have negative silhouette scores, meaning clusters may be overlapping or poorly separated.\n",
    "This is common in t-SNE/UMAP, as they optimize local structure rather than global clustering.\n",
    "\n",
    "AMI scores:\n",
    "Most AMI scores surpass 0.4, which indicates moderately good label-cluster alignment."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot kNN accuracy vs. recall for different dimensionality reduction methods\n",
    "\n",
    "def plot_knn_accuracy_vs_recall(results_df, figsize=(8, 6)):\n",
    "    \"\"\"\n",
    "    Parameters:\n",
    "        results_df (DataFrame): must contain columns 'Method', 'kNN Accuracy', and 'kNN Recall (avg)'\n",
    "        figsize (tuple): size of the plot (width, height)\n",
    "    \"\"\"\n",
    "    methods = results_df[\"Method\"]\n",
    "    accuracy = results_df[\"kNN Accuracy\"]\n",
    "    recall = results_df[\"kNN Recall (avg)\"]\n",
    "\n",
    "    plt.figure(figsize=figsize)\n",
    "    texts = []\n",
    "\n",
    "    # plot points\n",
    "    for i in range(len(methods)):\n",
    "        plt.scatter(accuracy[i], recall[i], s=200, label=methods[i])\n",
    "        plt.text(accuracy[i], recall[i], methods[i], fontsize=8, rotation=15)\n",
    "\n",
    "    plt.xlabel(\"kNN Accuracy\")\n",
    "    plt.ylabel(\"kNN Recall (macro avg)\")\n",
    "    plt.title(\"kNN Accuracy vs. Recall for Dimensionality Reduction Methods\")\n",
    "    plt.grid(True)\n",
    "    plt.tight_layout()\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_knn_accuracy_vs_recall(results_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Interpretation:\n",
    "\n",
    "* t-SNE (2D) nearly matches the accuracy of the high-dimensional data while even slightly improving average recall. This shows that it seems to preserve neighborhood structure quite well in 2D.\n",
    "* UMAP (2D) performs reasonably, a bit behind t-SNE, but still far ahead of PCA (2D).\n",
    "* PCA (2D) performs poorly — this confirms PCA doesn't capture non-linear relationships or local structure well in low dimensions.\n",
    "* t-SNE and UMAP seem to have benefitted from PCA initialization => visualize these again below\n",
    "* High-dimensional kNN is very strong (as expected), but not better than t-SNE, UMAP or especially any of the other high-dimensional methods. Consequently, e.g. even the 2D t-SNE visualization seems to be biologically meaningful."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# visualize the 2D \"winners\"\n",
    "# define 2D projections\n",
    "X_pca2 = PCA(n_components=2).fit_transform(X_hvg)\n",
    "\n",
    "# precompute PCA-reduced input for non-linear methods\n",
    "X_pca_for_embedding = PCA(n_components=50).fit_transform(X_hvg)\n",
    "\n",
    "# t-SNE and UMAP from PCA\n",
    "X_tsne2 = TSNE(\n",
    "    n_components=2, perplexity=30, method=\"exact\", random_state=42\n",
    ").fit_transform(X_pca_for_embedding)\n",
    "X_umap2 = umap.UMAP(\n",
    "    n_components=2, n_neighbors=15, min_dist=0.1, random_state=42\n",
    ").fit_transform(X_pca_for_embedding)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot models from above\n",
    "fig, axes = plt.subplots(1, 3, figsize=(15, 5))\n",
    "\n",
    "titles = [\"PCA (2D)\", \"t-SNE (2D from PCA)\", \"UMAP (2D from PCA)\"]\n",
    "embeddings = [X_pca2, X_tsne2, X_umap2]\n",
    "\n",
    "for ax, title, embed in zip(axes, titles, embeddings):\n",
    "    sns.scatterplot(\n",
    "        x=embed[:, 0],\n",
    "        y=embed[:, 1],\n",
    "        hue=rna_family,\n",
    "        palette=\"tab10\",\n",
    "        s=30,\n",
    "        alpha=0.8,\n",
    "        legend=False,\n",
    "        ax=ax,\n",
    "    )\n",
    "    ax.set_title(title)\n",
    "    ax.set_xlabel(\"Dim 1\")\n",
    "    ax.set_ylabel(\"Dim 2\")\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Confirming the quantitative comparison, s-SNE and UMAP with PCA initialization look better than without, showing clearer clusters."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Task 4"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4.1 prepare data\n",
    "\n",
    "Outline of clustering steps was mainly inspired by Harris et al. 2018"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# as Harris et al. suggest:\n",
    "# use either raw count data or counts normalized per cell by size factors\n",
    "\n",
    "# we already have them:\n",
    "# log_cpm_df --> log-transformed CPM-normalized data\n",
    "# exonCounts --> raw counts\n",
    "# cpm --> CPM-normalized data\n",
    "# total_counts_per_cell --> total counts per cell\n",
    "# hvg_genes --> highly variable gene names\n",
    "# For leiden, we also need: log_hvg_ df (= log_cpm_df[hvg_genes])\n",
    "\n",
    "# for GMM wen can (optionally) standardize the data\n",
    "X_log_hvg_scaled = StandardScaler().fit_transform(log_hvg_df.values)\n",
    "\n",
    "# for NBMM, we will need\n",
    "exonCounts_df = pd.DataFrame(exonCounts, columns=genes)\n",
    "raw_hvg_df = exonCounts_df[hvg_genes_500]\n",
    "\n",
    "# and normalize by \"library size\", i.e. per-cell size factors\n",
    "size_factors = raw_hvg_df.sum(axis=1) / np.median(raw_hvg_df.sum(axis=1))\n",
    "normalized_counts_nbmm = raw_hvg_df.div(size_factors, axis=0)\n",
    "\n",
    "# Compute size factors (total counts per cell)\n",
    "# Rescale entire matrix to a stable magnitude (~CPM scale)\n",
    "scaling_factor = 1e3 / normalized_counts_nbmm.values.mean()  # bring mean ~1000\n",
    "normalized_counts_nbmm *= scaling_factor\n",
    "\n",
    "print(\"log_hvg_df shape:\", log_hvg_df.shape)\n",
    "print(\"raw_hvg_df shape:\", raw_hvg_df.shape)\n",
    "print(\"normalized_counts_nbmm shape:\", normalized_counts_nbmm.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4.2 Leiden clustering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Input data\n",
    "X = X_log_hvg_scaled\n",
    "pca = PCA(n_components=25)\n",
    "X_pca = pca.fit_transform(X)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Store results\n",
    "results = []\n",
    "ari_type_leiden = []\n",
    "ari_family_leiden = []\n",
    "K_leiden = []\n",
    "\n",
    "n_neighbors_list = [15, 20, 30, 50]\n",
    "resolution_list = [0.5, 1.0, 1.5, 2.0]\n",
    "\n",
    "for n_neighbors in n_neighbors_list:\n",
    "    # kNN graph\n",
    "    knn = NearestNeighbors(n_neighbors=n_neighbors, metric=\"euclidean\").fit(\n",
    "        X_pca\n",
    "    )\n",
    "    distances, indices = knn.kneighbors(X_pca)\n",
    "\n",
    "    # Create edges from kNN indices\n",
    "    edges = []\n",
    "    for i, neighbors in enumerate(indices):\n",
    "        for neighbor in neighbors[1:]:  # skip self-loop\n",
    "            edges.append((i, neighbor))\n",
    "\n",
    "    # Build igraph graph\n",
    "    g = ig.Graph(edges=edges, directed=False)\n",
    "    g.simplify()\n",
    "\n",
    "    for resolution in resolution_list:\n",
    "        # Run Leiden clustering\n",
    "        partition = leidenalg.find_partition(\n",
    "            g,\n",
    "            leidenalg.RBConfigurationVertexPartition,\n",
    "            resolution_parameter=resolution,\n",
    "        )\n",
    "        leiden_labels = np.array(partition.membership)\n",
    "\n",
    "        # Compute ARIs\n",
    "        assert (\n",
    "            len(leiden_labels) == X.shape[0]\n",
    "        ), \"Clustering output doesn't match data\"\n",
    "        assert len(rna_type_filtered) >= X.shape[0], \"Too few labels!\"\n",
    "\n",
    "        # Align label arrays to X\n",
    "        # Slice to match number of cells in X / clustering output\n",
    "        rna_type_subset = np.array(rna_type_filtered)[: len(leiden_labels)]\n",
    "        rna_family_subset = np.array(rna_family_filtered)[: len(leiden_labels)]\n",
    "\n",
    "        ari_type_score = adjusted_rand_score(rna_type_subset, leiden_labels)\n",
    "        ari_family_score = adjusted_rand_score(\n",
    "            rna_family_subset, leiden_labels\n",
    "        )\n",
    "\n",
    "        # store for plotting\n",
    "        K = len(np.unique(leiden_labels))\n",
    "        K_leiden.append(K)\n",
    "        ari_type_leiden.append(ari_type_score)\n",
    "        ari_family_leiden.append(ari_family_score)\n",
    "\n",
    "        # Store result\n",
    "        results.append(\n",
    "            {\n",
    "                \"n_neighbors\": n_neighbors,\n",
    "                \"resolution\": resolution,\n",
    "                \"n_clusters\": len(np.unique(leiden_labels)),\n",
    "                \"ARI (RNA Type)\": ari_type_score,\n",
    "                \"ARI (RNA Family)\": ari_family_score,\n",
    "            }\n",
    "        )\n",
    "\n",
    "# Convert to DataFrame\n",
    "results_df = pd.DataFrame(results)\n",
    "display(results_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Quick note: \n",
    "\n",
    "→ Adjusted Rand Index (ARI) is used as a clustering quality measure because it quantifies the similarity between predicted clusters and true labels while correcting for chance, making it robust and interpretable for comparing clustering performance.\n",
    "\n",
    "*Interpretation:*\n",
    "\n",
    "* Increasing the resolution parameter generally increases the number of clusters.\n",
    "    * e.g. n_neighbors= 15, clusters from 7 (res 0.5) to 17 (res 2)\n",
    "* highest ARI for RNA_family is 0.528 at\n",
    "    * n_neighbors = 15\n",
    "    * resolution = 1.0\n",
    "    * n_clusters = 9 (matches the number of RNA families) → This is likely the most biologically meaningful Leiden clustering.  Provides strong evidence that the clustering is capturing biologically meaningful structure. An ARI of ~0.52 with a biologically matched number of clusters (9 RNA families) is strong support that the preprocessing and clustering choices were well-justified\n",
    "* trade-off in resolution\n",
    "    * Higher resolutions increase granularity, but ARI doesn’t necessarily improve\n",
    "    * In some cases (e.g., res = 2.0), more clusters lead to lower ARI, suggesting over-clustering may obscure true biological structure\n",
    "* RNA Types vs. RNA Family\n",
    "    * ARI (RNA Family) scores are consistently higher than ARI (RNA Type), across the board.\n",
    "    * → This suggests the clustering aligns more closely with RNA families than types — possibly because families are broader or better captured by the feature space.\n",
    "* Stability across neighbors\n",
    "    * The top ARI scores for RNA family are relatively consistent across different n_neighbors, especially at res=1.0 and 1.5.\n",
    "    * This suggests that the Leiden results are not overly sensitive to this parameter in our case.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot results\n",
    "fig, axes = plt.subplots(1, 2, figsize=(14, 5), sharey=True)\n",
    "\n",
    "sns.lineplot(\n",
    "    data=results_df,\n",
    "    x=\"resolution\",\n",
    "    y=\"ARI (RNA Type)\",\n",
    "    hue=\"n_neighbors\",\n",
    "    marker=\"o\",\n",
    "    ax=axes[0],\n",
    ")\n",
    "axes[0].set_title(\"ARI vs Resolution (RNA Type)\")\n",
    "axes[0].set_ylabel(\"Adjusted Rand Index\")\n",
    "axes[0].set_xlabel(\"Leiden Resolution\")\n",
    "\n",
    "sns.lineplot(\n",
    "    data=results_df,\n",
    "    x=\"resolution\",\n",
    "    y=\"ARI (RNA Family)\",\n",
    "    hue=\"n_neighbors\",\n",
    "    marker=\"o\",\n",
    "    ax=axes[1],\n",
    ")\n",
    "axes[1].set_title(\"ARI vs Resolution (RNA Family)\")\n",
    "axes[1].set_ylabel(\"Adjusted Rand Index\")\n",
    "axes[1].set_xlabel(\"Leiden Resolution\")\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*Interpretation:*\n",
    "\n",
    "The plot generally supports the numbers seen and described above. \n",
    "\n",
    "Left - RNA Type:\n",
    "\n",
    "* ARI increases consistently with resolution, regardless of neighbor count.\n",
    "* Lower neighbor counts (15, 20) tend to produce slightly higher ARI for RNA Type.\n",
    "* However, overall ARIs stay < 0.25, meaning RNA Type is less distinctly separated in the data.\n",
    "* RNA Type labels reflect less distinct expression profiles or more gradual transitions between groups. High resolution helps slightly, but separation is still weak.\n",
    "\n",
    "Right - RNA Family:\n",
    "\n",
    "* A clear performance peak at resolution = 1.0, especially for 15 neighbors.\n",
    "* ARIs decline for higher resolutions, likely due to over-segmentation (splitting coherent RNA families).\n",
    "* Best score (~0.52) occurs at resolution 1.0, 15 neighbors → matches biological ground truth of 9 families\n",
    "* Leiden clustering captures biologically relevant structure best at moderate resolution and tighter neighborhood definitions. Pushing to higher resolutions divides the RNA families into smaller, less meaningful clusters\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4.3 Gaussian Mixture Model\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Range of cluster numbers to try\n",
    "cluster_range = range(5, 21)  # try 5 to 20 clusters\n",
    "\n",
    "# Store results\n",
    "gmm_results = []\n",
    "\n",
    "for n_clusters in cluster_range:\n",
    "    gmm = GaussianMixture(n_components=n_clusters, random_state=42)\n",
    "    gmm_labels = gmm.fit_predict(X_log_hvg_scaled)\n",
    "\n",
    "    ari_type_gmm = adjusted_rand_score(rna_type, gmm_labels)\n",
    "    ari_family_gmm = adjusted_rand_score(rna_family, gmm_labels)\n",
    "\n",
    "    gmm_results.append(\n",
    "        {\n",
    "            \"n_clusters\": n_clusters,\n",
    "            \"ari_rna_type\": ari_type_gmm,\n",
    "            \"ari_rna_family\": ari_family_gmm,\n",
    "        }\n",
    "    )\n",
    "\n",
    "# Convert to DataFrame\n",
    "df_gmm_results = pd.DataFrame(gmm_results)\n",
    "\n",
    "# Display\n",
    "\n",
    "plt.figure(figsize=(8, 5))\n",
    "plt.plot(\n",
    "    df_gmm_results[\"n_clusters\"],\n",
    "    df_gmm_results[\"ari_rna_type\"],\n",
    "    label=\"ARI vs RNA Type\",\n",
    "    marker=\"o\",\n",
    ")\n",
    "plt.plot(\n",
    "    df_gmm_results[\"n_clusters\"],\n",
    "    df_gmm_results[\"ari_rna_family\"],\n",
    "    label=\"ARI vs RNA Family\",\n",
    "    marker=\"s\",\n",
    ")\n",
    "plt.xlabel(\"Number of GMM Clusters\")\n",
    "plt.ylabel(\"Adjusted Rand Index (ARI)\")\n",
    "plt.title(\"GMM Clustering: ARI vs Number of Clusters\")\n",
    "plt.grid(True)\n",
    "plt.legend()\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*Interpretation:*\n",
    "\n",
    "RNA Family (orange):\n",
    "\n",
    "* reaches highest ARI (~ 0.45-0.46) at K = 8-10, again aligning well with the true number of RNA families (9)\n",
    "* ARI drops sharply after K > 10, indicating over-segmentation\n",
    "* peak is slightly lower than Leiden’s best (~0.52)\n",
    "* GMM captures meaningful structure for RNA families when the cluster count is close to ground truth. Like Leiden, it performs best when K ≈ number of RNA families. However, it is slightly less expressive or flexible than Leiden in this task\n",
    "\n",
    "\n",
    "RNA Type (blue):\n",
    "\n",
    "* ARIs increase steadily with K and then plateau around 0.16–0.17 from K = 14 onward.\n",
    "* Overall scores are modest, again suggesting that RNA type is not strongly clustered.\n",
    "* This trend matches the Leiden results: RNA type differences are subtle and not easily captured by unsupervised clustering, even in GMM which assumes Gaussian structure.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4.4 Negative Binomial Mixture Model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initialization via kmeans\n",
    "def initialize_responsibilities_kmeans(X, n_clusters):\n",
    "    kmeans = KMeans(n_clusters=n_clusters, n_init=5, random_state=42).fit(X)\n",
    "    labels_init = kmeans.labels_\n",
    "    resp = np.zeros((X.shape[0], n_clusters))\n",
    "    resp[np.arange(X.shape[0]), labels_init] = 1\n",
    "    return resp\n",
    "\n",
    "\n",
    "# put NBMM copmputation in a function\n",
    "def run_nbmm(X, K, r=2, max_iter=100, tol=1e-4, seed=42, verbose=False):\n",
    "    np.random.seed(seed)\n",
    "    N, G = X.shape\n",
    "\n",
    "    # Initialize with kmeans (see above)\n",
    "    resp = initialize_responsibilities_kmeans(X, K)\n",
    "    Z = np.argmax(resp, axis=1)\n",
    "\n",
    "    # Initial parameters\n",
    "    pi = resp.sum(axis=0) / N\n",
    "    p = np.zeros((K, G))\n",
    "\n",
    "    for k in range(K):\n",
    "        X_k = X[Z == k]\n",
    "        if X_k.shape[0] == 0:\n",
    "            X_k = X[np.random.choice(N, size=5, replace=False)]\n",
    "        mu_k = np.maximum(X_k.mean(axis=0), 1e-2)\n",
    "        p[k] = r / (r + mu_k)\n",
    "\n",
    "    prev_log_likelihood = -np.inf\n",
    "\n",
    "    for iteration in range(max_iter):\n",
    "        log_resp = np.zeros((N, K))\n",
    "\n",
    "        for k in range(K):\n",
    "            with np.errstate(divide=\"ignore\", invalid=\"ignore\"):\n",
    "\n",
    "                p_k = np.clip(p[k], 1e-5, 1 - 1e-5)\n",
    "                log_prob = nbinom.logpmf(X, n=r, p=p_k)\n",
    "                if not np.isfinite(log_prob).all():\n",
    "                    print(f\"Cluster {k}: log_prob contains NaN or -inf\")\n",
    "                log_resp[:, k] = np.sum(log_prob, axis=1) + np.log(\n",
    "                    pi[k] + 1e-8\n",
    "                )\n",
    "\n",
    "        # log-sum-exp stability\n",
    "        log_resp -= np.max(log_resp, axis=1, keepdims=True)\n",
    "        resp = np.exp(log_resp)\n",
    "        resp_sum = resp.sum(axis=1, keepdims=True)\n",
    "        resp_sum[resp_sum == 0] = 1e-8\n",
    "        resp /= resp_sum\n",
    "\n",
    "        # M-step\n",
    "        Nk = resp.sum(axis=0)\n",
    "        pi = Nk / N\n",
    "        for k in range(K):\n",
    "            weighted_sum = np.dot(resp[:, k], X)\n",
    "            mean_k = weighted_sum / (Nk[k] + 1e-8)\n",
    "            p[k] = r / (r + np.maximum(mean_k, 1e-2))\n",
    "\n",
    "        # Convergence\n",
    "        current_log_likelihood = np.sum(np.log(resp_sum))\n",
    "        if verbose:\n",
    "            print(\n",
    "                f\"Iter {iteration}, Log Likelihood: {current_log_likelihood:.2f}\"\n",
    "            )\n",
    "\n",
    "        if np.abs(current_log_likelihood - prev_log_likelihood) < tol:\n",
    "            break\n",
    "        prev_log_likelihood = current_log_likelihood\n",
    "\n",
    "    labels = np.argmax(resp, axis=1)\n",
    "    return labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# X_nbmm = normalized_counts_nbmm.values\n",
    "X_nbmm = np.round(raw_hvg_df.values).astype(int)\n",
    "rna_type_array = np.array(rna_type_filtered)\n",
    "rna_family_array = np.array(rna_family_filtered)\n",
    "\n",
    "K_values = range(5, 21)\n",
    "ari_type_nbmm = []\n",
    "ari_family_nbmm = []\n",
    "\n",
    "for K in K_values:\n",
    "    print(f\"Running NBMM for K={K}\")\n",
    "    labels = run_nbmm(X_nbmm, K, r=2)\n",
    "    ari_type_nbmm.append(adjusted_rand_score(rna_type_array, labels))\n",
    "    ari_family_nbmm.append(adjusted_rand_score(rna_family_array, labels))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure(figsize=(10, 6))\n",
    "plt.plot(K_values, ari_type_nbmm, label=\"ARI vs RNA Type\", marker=\"o\")\n",
    "plt.plot(K_values, ari_family_nbmm, label=\"ARI vs RNA Family\", marker=\"s\")\n",
    "plt.xlabel(\"Number of Clusters (K)\")\n",
    "plt.ylabel(\"Adjusted Rand Index\")\n",
    "plt.title(\"NBMM Clustering Performance\")\n",
    "plt.legend()\n",
    "plt.grid(True)\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*Interpretation:*\n",
    "\n",
    "RNA Family (orange):\n",
    "\n",
    "* ARI scores peak at K = 10 (~0.32) and again around K = 12.\n",
    "* This is lower than Leiden (~0.52) and GMM (~0.46).\n",
    "* Still shows a recognizable signal for RNA family structure, but less sharp and stable.\n",
    "* NBMM captures RNA family structure to *some* extent, but it seems less expressive than Leiden or GMM here. This might be due to:\n",
    "    * Model assumptions (NBMM assumes overdispersed count data).\n",
    "    * Sensitivity to initialization or data preprocessing (e.g., normalization, rounding).\n",
    "\n",
    "RNA Type (blue):\n",
    "\n",
    "* ARI scores stay consistently low (0.04–0.11), much like with other methods.\n",
    "* Slight improvement with increasing K, but still limited.\n",
    "* Again, RNA type structure is weakly represented in the data, and NBMM (as the other methods) struggles to reveal it."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Leiden results come from varying n_neighbours and resolution, so k_leiden has duplicate values.\n",
    "\n",
    "Therefore, we will not plot all results but only the maximum ARI score for each k value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract GMM results\n",
    "K_gmm = df_gmm_results[\"n_clusters\"].values\n",
    "ari_type_gmm = df_gmm_results[\"ari_rna_type\"].values\n",
    "ari_family_gmm = df_gmm_results[\"ari_rna_family\"].values\n",
    "\n",
    "# Aggregate by number of clusters\n",
    "leiden_df = pd.DataFrame(\n",
    "    {\n",
    "        \"K\": K_leiden,\n",
    "        \"ARI_RNA_Type\": ari_type_leiden,\n",
    "        \"ARI_RNA_Family\": ari_family_leiden,\n",
    "    }\n",
    ")\n",
    "\n",
    "# Take maximum ARI per K\n",
    "leiden_summary = (\n",
    "    leiden_df.groupby(\"K\")\n",
    "    .agg({\"ARI_RNA_Type\": \"max\", \"ARI_RNA_Family\": \"max\"})\n",
    "    .reset_index()\n",
    ")\n",
    "\n",
    "plt.figure(figsize=(12, 6))\n",
    "\n",
    "# RNA Type\n",
    "plt.plot(\n",
    "    leiden_summary[\"K\"],\n",
    "    leiden_summary[\"ARI_RNA_Type\"],\n",
    "    \"o-\",\n",
    "    label=\"Leiden (RNA Type)\",\n",
    ")\n",
    "plt.plot(\n",
    "    df_gmm_results[\"n_clusters\"],\n",
    "    df_gmm_results[\"ari_rna_type\"],\n",
    "    \"s-\",\n",
    "    label=\"GMM (RNA Type)\",\n",
    ")\n",
    "plt.plot(K_values, ari_type_nbmm, \"^-\", label=\"NBMM (RNA Type)\")\n",
    "\n",
    "# RNA Family\n",
    "plt.plot(\n",
    "    leiden_summary[\"K\"],\n",
    "    leiden_summary[\"ARI_RNA_Family\"],\n",
    "    \"o--\",\n",
    "    label=\"Leiden (RNA Family)\",\n",
    ")\n",
    "plt.plot(\n",
    "    df_gmm_results[\"n_clusters\"],\n",
    "    df_gmm_results[\"ari_rna_family\"],\n",
    "    \"s--\",\n",
    "    label=\"GMM (RNA Family)\",\n",
    ")\n",
    "plt.plot(K_values, ari_family_nbmm, \"^--\", label=\"NBMM (RNA Family)\")\n",
    "\n",
    "plt.xlabel(\"Number of Clusters (K)\")\n",
    "plt.ylabel(\"Adjusted Rand Index (ARI)\")\n",
    "plt.title(\"Clustering Method Comparison by ARI (Max ARI per K for Leiden)\")\n",
    "plt.legend()\n",
    "plt.grid(True)\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Leiden clustering performs best overall, especially for RNA family labels. This indicates:\n",
    "\n",
    "* Graph-based clustering (Leiden) is more effective than GMM or NBMM in this context.\n",
    "* RNA family identity is more strongly reflected in expression data than RNA type.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# GMM\n",
    "best_idx_gmm_type = df_gmm_results[\"ari_rna_type\"].idxmax()\n",
    "best_idx_gmm_family = df_gmm_results[\"ari_rna_family\"].idxmax()\n",
    "\n",
    "# NBMM\n",
    "best_idx_nbmm_type = pd.Series(ari_type_nbmm).idxmax()\n",
    "best_idx_nbmm_family = pd.Series(ari_family_nbmm).idxmax()\n",
    "\n",
    "# Leiden\n",
    "best_idx_leiden_type = pd.Series(ari_type_leiden).idxmax()\n",
    "best_idx_leiden_family = pd.Series(ari_family_leiden).idxmax()\n",
    "\n",
    "# Build summary table\n",
    "summary_data = {\n",
    "    \"Method\": [\"GMM\", \"NBMM\", \"Leiden\"],\n",
    "    \"Best ARI (RNA Type)\": [\n",
    "        df_gmm_results[\"ari_rna_type\"].iloc[best_idx_gmm_type],\n",
    "        ari_type_nbmm[best_idx_nbmm_type],\n",
    "        ari_type_leiden[best_idx_leiden_type],\n",
    "    ],\n",
    "    \"Best K (RNA Type)\": [\n",
    "        df_gmm_results[\"n_clusters\"].iloc[best_idx_gmm_type],\n",
    "        K_values[best_idx_nbmm_type],\n",
    "        K_leiden[best_idx_leiden_type],\n",
    "    ],\n",
    "    \"Best ARI (RNA Family)\": [\n",
    "        df_gmm_results[\"ari_rna_family\"].iloc[best_idx_gmm_family],\n",
    "        ari_family_nbmm[best_idx_nbmm_family],\n",
    "        ari_family_leiden[best_idx_leiden_family],\n",
    "    ],\n",
    "    \"Best K (RNA Family)\": [\n",
    "        df_gmm_results[\"n_clusters\"].iloc[best_idx_gmm_family],\n",
    "        K_values[best_idx_nbmm_family],\n",
    "        K_leiden[best_idx_leiden_family],\n",
    "    ],\n",
    "}\n",
    "\n",
    "summary_df = pd.DataFrame(summary_data)\n",
    "display(summary_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run t-SNE on scaled log-HVG data\n",
    "tsne = TSNE(n_components=2, perplexity=30, n_iter=1000, random_state=42)\n",
    "X_tsne = tsne.fit_transform(X_pca)\n",
    "\n",
    "# Plot, colored by RNA Type\n",
    "plt.figure(figsize=(12, 5))\n",
    "\n",
    "plt.subplot(1, 2, 1)\n",
    "sns.scatterplot(\n",
    "    x=X_tsne[:, 0],\n",
    "    y=X_tsne[:, 1],\n",
    "    hue=np.array(rna_type_filtered)[: X_tsne.shape[0]],\n",
    "    palette=\"tab20\",\n",
    "    s=10,\n",
    "    alpha=0.7,\n",
    "    legend=False,\n",
    ")\n",
    "plt.title(\"t-SNE colored by RNA Type\")\n",
    "plt.xlabel(\"t-SNE 1\")\n",
    "plt.ylabel(\"t-SNE 2\")\n",
    "\n",
    "# Plot, colored by best Leiden clustering\n",
    "# Get best Leiden labels (highest ARI(RNA Family))\n",
    "best_index = np.argmax(ari_family_leiden)\n",
    "best_labels = results[best_index][\"n_clusters\"]\n",
    "\n",
    "# Recompute labels\n",
    "best_n_neighbors = results[best_index][\"n_neighbors\"]\n",
    "best_resolution = results[best_index][\"resolution\"]\n",
    "\n",
    "# Build graph again to get the best clustering\n",
    "knn = NearestNeighbors(n_neighbors=best_n_neighbors).fit(X_pca)\n",
    "_, indices = knn.kneighbors(X_pca)\n",
    "edges = [\n",
    "    (i, neighbor)\n",
    "    for i, neighbors in enumerate(indices)\n",
    "    for neighbor in neighbors[1:]\n",
    "]\n",
    "g = ig.Graph(edges=edges, directed=False)\n",
    "g.simplify()\n",
    "\n",
    "# Run Leiden clustering with the best resolution\n",
    "partition = leidenalg.find_partition(\n",
    "    g,\n",
    "    leidenalg.RBConfigurationVertexPartition,\n",
    "    resolution_parameter=best_resolution,\n",
    ")\n",
    "leiden_labels_best = np.array(partition.membership)\n",
    "\n",
    "# Plot\n",
    "plt.subplot(1, 2, 2)\n",
    "sns.scatterplot(\n",
    "    x=X_tsne[:, 0],\n",
    "    y=X_tsne[:, 1],\n",
    "    hue=leiden_labels_best,\n",
    "    palette=\"tab20\",\n",
    "    s=10,\n",
    "    alpha=0.7,\n",
    "    legend=False,\n",
    ")\n",
    "plt.title(\n",
    "    f\"t-SNE colored by Leiden Clusters\\n(Res={best_resolution}, Neigh={best_n_neighbors})\"\n",
    ")\n",
    "plt.xlabel(\"t-SNE 1\")\n",
    "plt.ylabel(\"t-SNE 2\")\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Leiden clustering works well visually; it’s producing meaningful structure that resembles real biological types.\n",
    "\n",
    "* The ARI score for RNA Family was highest for this configuration, and the right plot confirms that: there are 9 clear clusters, matching the number of RNA families.\n",
    "* The partial overlap seen in the RNA Type plot reflects biological or technical noise, and why clustering RNA type is harder (lower ARI).\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*Thoughts on NBMM performance:*\n",
    "\n",
    "Reasons why NBMM could perform so poorly:\n",
    "* Leiden and GMM use more expressive features\n",
    "    * Leiden uses neighborhood graphs on PCA-reduced, log-normalized data — it’s robust to noise and captures manifold structure.\n",
    "    * GMM works on scaled log-expression and handles ellipsoidal clusters well\n",
    "    * In contrast: NBMM operates directly on raw count data, which retain a lot of technical noise and sparsity (see task1)\n",
    "* Initialization & Optimization Challenges\n",
    "    * NBMM was initialized with KMeans and uses an EM-like algorithm, but this can:\n",
    "        * get stuck in local optima\n",
    "        * be slow to converge, especially on sparse, high-dimensional count matrices\n",
    "    * without careful parameter tuning (e.g. better initialization, r selection, regularization), it may not find meaningful clusters.\n",
    "* RNA Type (length: 77) vs. RNA Family (length: 9)\n",
    "    * Your ARI scores show NBMM performs better for RNA Family than for RNA Type.\n",
    "    * This suggests NBMM can capture coarse structure, but not finer subclass separations — likely due to limited flexibility or resolution in the model\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Task 5"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5.1 prepare data\n",
    "\n",
    "* transcriptomics data is already normalized, hvgs were selected, PCA applied\n",
    "* electrophysiological data: standardize/normalize (z-score), maybe apply PCA\n",
    "\n",
    "then:\n",
    "* match cells across modalities (=align them via cell IDs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "meta_df = meta.copy()\n",
    "meta_df = meta_df[\n",
    "    exclude_low_quality\n",
    "]  # filter out low-quality cells from meta\n",
    "meta_df[\"Cell\"] = meta_df[\"Cell\"].astype(str).str.strip()\n",
    "meta_df = meta_df.set_index(\"Cell\")\n",
    "\n",
    "# set the same index on RNA and ephys\n",
    "log_cpm_df.index = meta_df.index[: log_cpm_df.shape[0]]\n",
    "ephys_df.index = meta_df.index[: ephys_df.shape[0]]\n",
    "\n",
    "# z-score electrophysiological data\n",
    "scaler = StandardScaler()\n",
    "ephys_df_scaled = pd.DataFrame(\n",
    "    scaler.fit_transform(ephys_df),\n",
    "    index=ephys_df.index,\n",
    "    columns=ephys_df.columns,\n",
    ")\n",
    "ephys_df = ephys_df_scaled\n",
    "\n",
    "shared_cells = log_cpm_df.index.intersection(ephys_df.index).intersection(\n",
    "    meta_df.index\n",
    ")\n",
    "print(\"Number of shared cells:\", len(shared_cells))\n",
    "\n",
    "log_cpm_df = log_cpm_df.loc[shared_cells]\n",
    "ephys_df = ephys_df.loc[shared_cells]\n",
    "meta_df = meta_df.loc[shared_cells]\n",
    "\n",
    "# double check that alignment is correct\n",
    "assert all(log_cpm_df.index == ephys_df.index)\n",
    "assert all(ephys_df.index == meta_df.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"Final aligned shapes:\")\n",
    "print(\"RNA:\", log_cpm_df.shape)\n",
    "print(\"Ephys:\", ephys_df.shape)\n",
    "print(\"Meta:\", meta_df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# recompute HVGs\n",
    "\n",
    "# use variance thresholding to mimic HVG selection\n",
    "selector = VarianceThreshold(threshold=0.5)  # adjusted threshold\n",
    "_ = selector.fit(log_cpm_df)\n",
    "\n",
    "hvg_genes = log_cpm_df.columns[selector.get_support()]\n",
    "print(f\"Selected {len(hvg_genes)} HVGs based on current log_cpm_df.\")\n",
    "\n",
    "# rerun PCA\n",
    "X_hvg = StandardScaler().fit_transform(log_cpm_df[hvg_genes])\n",
    "\n",
    "pca = PCA(n_components=25)  # chose to keep 25 PCs\n",
    "\n",
    "X_pca = pca.fit_transform(X_hvg)\n",
    "\n",
    "# convert to df for easier correlation\n",
    "pca_df = pd.DataFrame(\n",
    "    X_pca,\n",
    "    columns=[f\"PC{i+1}\" for i in range(X_pca.shape[1])],\n",
    "    index=log_cpm_df.index,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "now: explore relationships\n",
    "\n",
    "* correlation analysis\n",
    "* canonical correlation analysis\n",
    "* UMAP\n",
    "* PCA loadings to see gene contributions\n",
    "* regression models\n",
    "* biological interpretation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5.2 correlation analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for all features vs all PCs:\n",
    "cor_matrix_full = pd.DataFrame(index=ephys_df.columns, columns=pca_df.columns)\n",
    "for ef in ephys_df.columns:\n",
    "    for pc in pca_df.columns:\n",
    "        cor_matrix_full.loc[ef, pc] = ephys_df[ef].corr(pca_df[pc])\n",
    "cor_matrix_full = cor_matrix_full.astype(float)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(12, 8))\n",
    "sns.heatmap(\n",
    "    cor_matrix_full,\n",
    "    annot=True,\n",
    "    cmap=\"vlag\",\n",
    "    center=0,\n",
    "    annot_kws={\"size\": 7},\n",
    "    ax=ax,\n",
    ")\n",
    "ax.set_title(\n",
    "    \"Correlation between electrophysiological features and PCA components\"\n",
    ")\n",
    "ax.set_xlabel(\"Principal Components\")\n",
    "ax.set_ylabel(\"Ephys Features\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The heatmap shows Pearson correlation coefficients between the electrophysiological features and the transcriptomic PCs. We can see some negative values, indicating inverse correlation:\n",
    "* Upstroke-to-downstroke ratio is negatively correlated with PC3, which suggests that higher values on PC3 go along with lower upstroke-downstroke ratios\n",
    "* AP width and Afterhyperpolarization are also negatively correlated with PC3\n",
    "\n",
    "There are also some positive values indicating positive correlation:\n",
    "* higher Input resistance seems to go along with a high score for PC4\n",
    "* higher Max number of APs seems to go along with a high score for PC3\n",
    "\n",
    "Also, there are many near-zero values (white) that suggest no linear relationship between both types of data. Overall, PC3 stands out with the strongest ephys-correlation."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we want to isolate ion channel gene expression from data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert all column names to uppercase to standardize for matching\n",
    "log_cpm_df.columns = log_cpm_df.columns.str.upper()\n",
    "\n",
    "# choose channel gene families of interest\n",
    "ion_channel_prefixes = [\"KCN\", \"SCN\", \"HCN\", \"CACNA\", \"KCNA\", \"GRIN\", \"GABR\"]\n",
    "\n",
    "# select genes that contain any of those chosen ion channel prefixes\n",
    "ion_genes = [\n",
    "    gene\n",
    "    for gene in log_cpm_df.columns\n",
    "    if any(prefix in gene for prefix in ion_channel_prefixes)\n",
    "]\n",
    "\n",
    "# subset expression matrix to only those genes\n",
    "ion_expr = log_cpm_df[ion_genes]\n",
    "\n",
    "print(f\"Found {len(ion_genes)} ion channel genes in log_cpm_df.\")\n",
    "print(\"Top ion channel genes found:\", ion_genes[:10])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# double check that index alignment is correct\n",
    "common_cells = ion_expr.index.intersection(ephys_df.index)\n",
    "ion_expr = ion_expr.loc[common_cells]\n",
    "ephys_subset = ephys_df.loc[common_cells][\n",
    "    [\"AP threshold (mV)\", \"Rheobase (pA)\", \"Membrane time constant (ms)\"]\n",
    "]\n",
    "# check for same order\n",
    "assert all(ion_expr.index == ephys_subset.index), \"index order mismatch\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "After aligning them, we can now correlate each ion gene with each ephys feature:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# correlation matrix between genes and selected ephys features\n",
    "cor_matrix = ion_expr.corrwith(\n",
    "    ephys_subset[\"AP threshold (mV)\"], axis=0\n",
    ").to_frame(name=\"AP threshold (mV)\")\n",
    "\n",
    "cor_matrix[\"Rheobase (pA)\"] = ion_expr.corrwith(\n",
    "    ephys_subset[\"Rheobase (pA)\"], axis=0\n",
    ")\n",
    "\n",
    "cor_matrix[\"Membrane time constant (ms)\"] = ion_expr.corrwith(\n",
    "    ephys_subset[\"Membrane time constant (ms)\"], axis=0\n",
    ")\n",
    "\n",
    "# make matrix easier to read + ensure numeric\n",
    "cor_matrix = cor_matrix.T.astype(float)\n",
    "\n",
    "# sort genes by correlation strength to Rheobase => makes the plot easier to read\n",
    "sorted_genes = (\n",
    "    cor_matrix.loc[\"Rheobase (pA)\"].abs().sort_values(ascending=False).index\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# visualize the correlations:\n",
    "plt.figure(figsize=(12, 6))\n",
    "sns.heatmap(\n",
    "    cor_matrix[sorted_genes],\n",
    "    annot=None,\n",
    "    cmap=\"vlag\",\n",
    "    center=0,\n",
    "    fmt=\"\",\n",
    ")\n",
    "plt.title(\"Correlation between Ion Channel Gene Expression and Ephys Features\")\n",
    "plt.xlabel(\"Ion Channel Genes\")\n",
    "plt.ylabel(\"Ephys Features\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*Interpretation*\n",
    "\n",
    "* Rheobase shows the stronges correlations with many genes, which means that expression of these ion channel genes increases with Rheobase => more current is needed to elicit spikes. Genes included here are for example GRIN2A, KCNMA1 and SCN4B\n",
    "* KCNK family genes (potassium channels) correlate with membrane time constant and AP threshold, at least to a medium degree. This is expected, since they play a known role in setting resting potential and excitability.\n",
    "* Some SCN genes (sodium channels) also show rather small correlations, especially with AP threshold and Rheobase, which is also expected, since it is consistent with their role in spike initiation.\n",
    "\n",
    "Overall, most correlations are pretty weak => check for significance!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# compute correlations & p-values\n",
    "results = []\n",
    "for gene in ion_expr.columns:\n",
    "    for feat in ephys_subset.columns:\n",
    "        r, p = pearsonr(ion_expr[gene], ephys_subset[feat])\n",
    "        results.append({\"gene\": gene, \"feature\": feat, \"r\": r, \"p\": p})\n",
    "\n",
    "cor_df = pd.DataFrame(results)\n",
    "\n",
    "# FDR correction\n",
    "cor_df[\"fdr_pass\"], cor_df[\"qval\"] = fdrcorrection(cor_df[\"p\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# reuse correlation matrix from before\n",
    "# this time add significance stars\n",
    "annot_matrix = pd.DataFrame(index=cor_matrix.index, columns=cor_matrix.columns)\n",
    "\n",
    "r_matrix = cor_df.pivot(index=\"feature\", columns=\"gene\", values=\"r\")\n",
    "qval_matrix = cor_df.pivot(index=\"feature\", columns=\"gene\", values=\"qval\")\n",
    "\n",
    "\n",
    "for row in cor_matrix.index:\n",
    "    for col in cor_matrix.columns:\n",
    "        r_val = cor_matrix.at[row, col]\n",
    "        q_val = qval_matrix.at[\n",
    "            row, col\n",
    "        ]\n",
    "\n",
    "        if pd.isna(r_val) or pd.isna(q_val):\n",
    "            annot_matrix.at[row, col] = \"\"\n",
    "            continue\n",
    "\n",
    "        stars = \"\"\n",
    "        if q_val < 0.001:\n",
    "            stars = \"*\\n*\\n*\"\n",
    "        elif q_val < 0.01:\n",
    "            stars = \"*\\n*\"\n",
    "        elif q_val < 0.05:\n",
    "            stars = \"*\"\n",
    "\n",
    "        annot_matrix.at[row, col] = stars\n",
    "\n",
    "# plot\n",
    "plt.figure(figsize=(16, 6))\n",
    "sns.heatmap(\n",
    "    cor_matrix[sorted_genes],  # sort by rheobase\n",
    "    annot=annot_matrix[sorted_genes],  # sorted annotation\n",
    "    fmt=\"\",\n",
    "    cmap=\"vlag\",\n",
    "    center=0,\n",
    "    annot_kws={\"size\": 10},\n",
    "    cbar_kws={\"label\": \"Pearson r\"},\n",
    ")\n",
    "plt.title(\n",
    "    \"Correlation Between Ion Channel Genes and Ephys Features\\n(* = FDR < 0.05)\"\n",
    ")\n",
    "plt.xlabel(\"Ion Channel Genes (sorted by Rheobase correlation)\")\n",
    "plt.ylabel(\"Electrophysiological Features\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This time, significant correlations are marked with 3, 2 or 1 stars, depending on the strength of the significance (q_val < 0.001 / 0.01 / 0.05)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To investigate the significant correlations further, look at the top 6 strongest ones:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# filter + select top 6 FDR-significant correlations\n",
    "top_hits = (\n",
    "    cor_df[cor_df[\"fdr_pass\"]]\n",
    "    .sort_values(\"r\", key=abs, ascending=False)\n",
    "    .head(6)\n",
    "    .reset_index(drop=True)\n",
    ")\n",
    "\n",
    "# plot scatterplots\n",
    "fig, axes = plt.subplots(2, 3, figsize=(18, 10))\n",
    "axes = axes.flatten()\n",
    "\n",
    "for i, row in top_hits.iterrows():\n",
    "    gene = row[\"gene\"]\n",
    "    feat = row[\"feature\"]\n",
    "    r_val = row[\"r\"]\n",
    "    q_val = row[\"qval\"]\n",
    "\n",
    "    x = ion_expr[gene]\n",
    "    y = ephys_subset[feat]\n",
    "\n",
    "    sns.scatterplot(x=x, y=y, ax=axes[i], alpha=0.6, edgecolor=None)\n",
    "    sns.regplot(x=x, y=y, ax=axes[i], scatter=False, color=\"red\", ci=None)\n",
    "\n",
    "    axes[i].set_title(\n",
    "        f\"{gene} vs. {feat}\\n\" f\"r = {r_val:.2f}, q = {q_val:.1e}\", fontsize=12\n",
    "    )\n",
    "    axes[i].set_xlabel(f\"{gene} expression (log CPM)\")\n",
    "    axes[i].set_ylabel(feat)\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.suptitle(\n",
    "    \"Top 6 FDR-Significant Gene–Ephys Correlations\", fontsize=16, y=1.02\n",
    ")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The plot above shows the top 6 statistically significant correlations between ion channel gene expression and ephys features. Numerically, the effect sizes are small. Also, the vertical streaks at 0 mean that a large fraction of cells had undetectable expression for the gene in question. This sparsity might weaken the correlations.\n",
    "However, the very low q-values (<3e-5) actually suggest pretty strong statistical significance.\n",
    "\n",
    "Interpretations of the single plots:\n",
    "\n",
    "* CACNA2D3 vs. Membrane Time Constant: CACNA2D3 encodes a subunit of voltage-gated calcium channels. Higher expression is associated with slower membrane dynamics (increased time constant), which might reflect more sustained calcium-mediated currents or associated modulation of channel density.\n",
    "* GRIN2A vs Rheobase: GRIN2A encodes an NMDA receptor subunit. Higher expression correlates with higher rheobase. This suggests that NMDA receptor signaling may contribute to increased excitability thresholds or delayed spiking, possibly through tonic depolarization or dendritic filtering.\n",
    "* CACNA1E vs. Membrane Time Constant: CACNA1E encodes the R-type Ca^2+ channel α1E subunit. Again, we see a positive correlation with the time constant, supporting a general link between Ca^2+ channel expression and slower membrane charging.\n",
    "* KCNH7/KCNH4 vs Rheobase: KCNH7 and KCNH4 are voltage-gated potassium channel. Our results suggest that more KCNH7 or KCNH4 expression is associated with greater current needed to fire, consistent with increased leak or repolarizing conductance reducing excitability.\n",
    "* KCNQ1OT1 vs Rheobase: KCNQ1OT1 is not a protein-coding channel itself, but a long non-coding RNA associated with KCNQ1. The foudn correlation might reflect regulation of KCNQ1 expression."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5.3 canonical correlation analysis\n",
    "\n",
    "=> inspired by Lause et al. (2024)\n",
    "\n",
    "Identify pairs of latent dimensions that co-vary across RNA and e-phys."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# standardize both views\n",
    "X_rna = StandardScaler().fit_transform(pca_df)\n",
    "X_ephys = StandardScaler().fit_transform(ephys_df)\n",
    "\n",
    "# apply CCA\n",
    "cca = CCA(n_components=5)\n",
    "X_rna_cca, X_ephys_cca = cca.fit_transform(X_rna, X_ephys)\n",
    "\n",
    "# visualize CCA components\n",
    "cca_df = pd.DataFrame(\n",
    "    np.concatenate([X_rna_cca, X_ephys_cca], axis=1),\n",
    "    columns=[f\"RNA_CCA{i+1}\" for i in range(X_rna_cca.shape[1])]\n",
    "    + [f\"EPHYS_CCA{i+1}\" for i in range(X_ephys_cca.shape[1])],\n",
    "    index=pca_df.index,\n",
    ")\n",
    "\n",
    "# correlation between CCA components\n",
    "corr_matrix = pd.DataFrame(\n",
    "    np.corrcoef(X_rna_cca.T, X_ephys_cca.T)[:5, 5:],\n",
    "    index=[f\"RNA_CCA{i+1}\" for i in range(5)],\n",
    "    columns=[f\"EPHYS_CCA{i+1}\" for i in range(5)],\n",
    ")\n",
    "\n",
    "\n",
    "\n",
    "sns.heatmap(corr_matrix, annot=True, cmap=\"vlag\", center=0)\n",
    "plt.title(\"Canonical Correlation between RNA and Ephys components\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The diagonal values in the above heatmap show the canonical correlation between matched components. The red values (around 0.5) suggest that there is a strong linear relationship between the respective genes and ephys traits. The off-diagonal values are very close to zero, which is good, since this suggests that each CCA component pair captures a distinct shared signal and not redundant info.\n",
    "\n",
    "We've seen that there transcriptomic and ephys data share structure => What genes and what electrophysiological features are responsible for this coupling?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# loadings = contribution of each original variable to the CCA components\n",
    "gene_loadings = pd.DataFrame(\n",
    "    cca.x_weights_,\n",
    "    index=pca_df.columns,  # original RNA PC names\n",
    "    columns=[f\"RNA_CCA{i+1}\" for i in range(cca.n_components)],\n",
    ")\n",
    "\n",
    "ephys_loadings = pd.DataFrame(\n",
    "    cca.y_weights_,\n",
    "    index=ephys_df.columns,\n",
    "    columns=[f\"EPHYS_CCA{i+1}\" for i in range(cca.n_components)],\n",
    ")\n",
    "\n",
    "# look at top contributing features to RNA_CCA1 and EPHYS_CCA1\n",
    "top_rna = gene_loadings[\"RNA_CCA1\"].abs().sort_values(ascending=False).head(10)\n",
    "top_ephys = (\n",
    "    ephys_loadings[\"EPHYS_CCA1\"].abs().sort_values(ascending=False).head(10)\n",
    ")\n",
    "\n",
    "print(\"Top genes (RNA CCA1):\\n\", top_rna, sep=\"\")\n",
    "print(\"-------------------------\", sep=\"\")\n",
    "print(\"Top ephys features (Ephys CCA1):\\n\", top_ephys, sep=\"\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Below, we visualize cell-level CCA projections, colored by Metadata. This is to see whether biological classes are reflected in shared gene–ephys structure."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "cca_proj = pd.DataFrame(\n",
    "    {\n",
    "        \"RNA_CCA1\": X_rna_cca[:, 0],\n",
    "        \"EPHYS_CCA1\": X_ephys_cca[:, 0],\n",
    "        \"Cre\": meta_df[\"Cre\"],\n",
    "    }\n",
    ")\n",
    "\n",
    "# custom palette for Cre lines\n",
    "unique_cres = cca_proj[\"Cre\"].unique()\n",
    "cre_palette = dict(\n",
    "    zip(unique_cres, sns.color_palette(\"tab20\", len(unique_cres)))\n",
    ")\n",
    "\n",
    "plt.figure(figsize=(10, 6))\n",
    "sns.scatterplot(\n",
    "    data=cca_proj,\n",
    "    x=\"RNA_CCA1\",\n",
    "    y=\"EPHYS_CCA1\",\n",
    "    hue=\"Cre\",\n",
    "    palette=cre_palette,\n",
    "    s=30,  # uniform point size\n",
    "    edgecolor=\"none\",  # cleaner look\n",
    "    alpha=0.7,\n",
    ")\n",
    "\n",
    "plt.axhline(0, color=\"gray\", linestyle=\"--\", lw=0.7)\n",
    "plt.axvline(0, color=\"gray\", linestyle=\"--\", lw=0.7)\n",
    "\n",
    "plt.title(\"CCA Projection of Cells (colored by Cre Line)\", fontsize=14)\n",
    "plt.xlabel(\"RNA CCA Component 1\", fontsize=12)\n",
    "plt.ylabel(\"Ephys CCA Component 1\", fontsize=12)\n",
    "\n",
    "# place legend outside the plot\n",
    "plt.legend(\n",
    "    title=\"Cre Line\",\n",
    "    bbox_to_anchor=(1.05, 1),\n",
    "    loc=\"upper left\",\n",
    "    borderaxespad=0.0,\n",
    "    fontsize=9,\n",
    "    title_fontsize=10,\n",
    ")\n",
    "\n",
    "plt.grid(True, linestyle=\"--\", alpha=0.3)\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The axes in the plot are the projections onto the first canonical components from CCA. Coloring was done based on Cre line identity, as drawn from metadata.\n",
    "\n",
    "The plot shows a diagonal trend, which confirms that there is a strong gene-ephys coupling. Also, even though not very clearly, some Cre lines cluster along specific parts of the diagonal. For example, PV-, SST+ and VIP+ occupy somewhat distinct regions, which shows their distinct gene-ephys signatures. This confirms that the canonical axes are biologically meaningful. To sum up, the first CCA component captures a shared axis of transcriptomic and electrophysiological variation that also aligns with genetic Cre-defined cell types."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5.4 UMAP over combined RNA PCs and scaled ephys features for joint manifold visualization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.preprocessing import StandardScaler\n",
    "import umap.umap_ as umap\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "\n",
    "meta_for_color = \"Cre\"  # \"RNA family\" / \"Slice\" / \"Sequencing batch\"\n",
    "\n",
    "# align indices across data\n",
    "shared_cells = pca_df.index.intersection(ephys_df.index).intersection(\n",
    "    meta_df.index\n",
    ")\n",
    "\n",
    "# combine RNA PCs + scaled ephys\n",
    "X_rna = pca_df.loc[shared_cells].values\n",
    "X_ephys = StandardScaler().fit_transform(ephys_df.loc[shared_cells])\n",
    "X_combined = np.concatenate([X_rna, X_ephys], axis=1)\n",
    "\n",
    "# UMAP\n",
    "reducer = umap.UMAP(\n",
    "    n_neighbors=15, min_dist=0.3, metric=\"euclidean\", random_state=42\n",
    ")\n",
    "embedding = reducer.fit_transform(X_combined)\n",
    "\n",
    "embedding_df = pd.DataFrame(\n",
    "    embedding, columns=[\"UMAP1\", \"UMAP2\"], index=shared_cells\n",
    ")\n",
    "embedding_df[\"ColorGroup\"] = meta_df.loc[shared_cells, meta_for_color]\n",
    "\n",
    "# plot\n",
    "plt.figure(figsize=(8, 6))\n",
    "sns.scatterplot(\n",
    "    data=embedding_df,\n",
    "    x=\"UMAP1\",\n",
    "    y=\"UMAP2\",\n",
    "    hue=\"ColorGroup\",\n",
    "    s=30,\n",
    "    alpha=0.7,\n",
    "    edgecolor=None,\n",
    "    palette=\"tab20\" if embedding_df[\"ColorGroup\"].nunique() <= 20 else \"husl\",\n",
    ")\n",
    "plt.axhline(0, color=\"gray\", linestyle=\"--\", lw=0.5)\n",
    "plt.axvline(0, color=\"gray\", linestyle=\"--\", lw=0.5)\n",
    "plt.title(\"Joint UMAP: RNA PCs + Electrophysiological Features\", fontsize=14)\n",
    "plt.legend(bbox_to_anchor=(1.05, 1), loc=\"upper left\", title=meta_for_color)\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*Interpretation:*\n",
    "\n",
    "Some Cre lines (e.g., PV+, VIP+) are more present in specific clusters, suggesting that these lines correspond to functionally and molecularly coherent cell types.\n",
    "Others (e.g., WT, VIPR2-) are more spread out, which implies broader variability or technical artifacts. Also, the cells aren't clustered purely by RNA or ephys. This shows that both modalities contribute to structure, which is the goal of joint UMAP."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5.5 study PCA loadings to see gene contributions\n",
    "\n",
    "=> look at what the PCs are made of"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "loadings = pd.DataFrame(\n",
    "    pca.components_.T,\n",
    "    index=hvg_genes,\n",
    "    columns=[f\"PC{i+1}\" for i in range(pca.n_components_)],\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "top_loadings = {}\n",
    "for pc in loadings.columns:\n",
    "    top = (\n",
    "        loadings[pc].abs().sort_values(ascending=False).head(10).index.tolist()\n",
    "    )\n",
    "    top_loadings[pc] = top"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "top_genes_pc1 = top_loadings[\"PC1\"]  # gene names\n",
    "\n",
    "# select just those genes' loadings across all PCs (or just PC1)\n",
    "sns.heatmap(\n",
    "    loadings.loc[top_genes_pc1],\n",
    "    cmap=\"coolwarm\",\n",
    "    annot=True,\n",
    "    fmt=\".2f\",\n",
    "    linewidths=0.5,\n",
    "    linecolor=\"gray\",\n",
    "    annot_kws={\"size\": 5},\n",
    "    cbar_kws={\"label\": \"Loading Value\"},\n",
    ")\n",
    "\n",
    "plt.title(\"Top Gene Loadings for PC1\")\n",
    "plt.xlabel(\"Principal Components\")\n",
    "plt.ylabel(\"Genes\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The cells in the plot represent how strongly (positively or negatively) each gene contributes to each PC. Look at the loadings in more detail:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# choose which PCs to visualize\n",
    "pcs_to_plot = [\"PC1\", \"PC2\", \"PC3\"]  # change or extend this list\n",
    "\n",
    "# set up subplots\n",
    "fig, axes = plt.subplots(\n",
    "    1, len(pcs_to_plot), figsize=(6 * len(pcs_to_plot), 5), sharey=False\n",
    ")\n",
    "\n",
    "if len(pcs_to_plot) == 1:\n",
    "    axes = [axes]  # ensure axes is always iterable\n",
    "\n",
    "for ax, pc in zip(axes, pcs_to_plot):\n",
    "    top_genes = loadings[pc].abs().sort_values(ascending=False).head(10).index\n",
    "    top_values = loadings.loc[top_genes, pc]\n",
    "\n",
    "    sns.barplot(\n",
    "        x=top_values.values,\n",
    "        y=top_genes,\n",
    "        hue=top_genes,\n",
    "        ax=ax,\n",
    "        palette=\"viridis\" if top_values.iloc[0] > 0 else \"rocket\",\n",
    "        orient=\"h\",\n",
    "        legend=False,\n",
    "    )\n",
    "    ax.set_title(f\"Top 10 Loadings for {pc}\")\n",
    "    ax.set_xlabel(\"Loading Value\")\n",
    "    ax.set_ylabel(\"Gene\")\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "PC1 (left panel):\n",
    "\n",
    "Genes like TRIM37, HDLPP, PCP4L1, DUSP10 are the top contributors.\n",
    "All values are positive, suggesting PC1 is dominated by a positively co-expressed gene module.\n",
    "If PC1 separates samples by cell type, region, or batch, these genes may underlie that variation.\n",
    "\n",
    "PC2 (middle panel):\n",
    "\n",
    "Mix of positive and negative loadings (e.g., UCK1: +0.07, RB1CC1: -0.06).\n",
    "Suggests PC2 contrasts two opposing gene sets: samples with high PC2 scores express UCK1 etc., and low PC2 scores express RB1CC1, PER2 etc.\n",
    "\n",
    "PC3 (right panel):\n",
    "\n",
    "Similar contrast between MTUS2/ARID5B (positive) and MAP2K4/VIPAS39 (negative).\n",
    "May represent a different biological axis (e.g., membrane vs. cytoskeletal genes => open for investigation)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5.6 regression models\n",
    "\n",
    "Predict e-phys features from transcriptomic data:\n",
    "\n",
    "* Use linear regression, random forests, and elastic net on PCA-reduced gene expression.\n",
    "* Evaluate using cross-validation (R^2, RMSE)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# features to predict\n",
    "ephys_targets = [\n",
    "    \"AP threshold (mV)\",\n",
    "    \"Rheobase (pA)\",\n",
    "    \"Membrane time constant (ms)\",\n",
    "]\n",
    "\n",
    "models = {\n",
    "    \"Linear Regression\": LinearRegression(),\n",
    "    \"Random Forest\": RandomForestRegressor(n_estimators=100, random_state=42),\n",
    "    \"Elastic Net\": ElasticNet(alpha=0.1, l1_ratio=0.5, random_state=42),\n",
    "}\n",
    "\n",
    "results = []\n",
    "\n",
    "# subplot grid: 3 ephys targets × 3 models\n",
    "fig, axes = plt.subplots(len(ephys_targets), len(models), figsize=(18, 12))\n",
    "axes = axes.reshape(len(ephys_targets), len(models))\n",
    "\n",
    "for i, target in enumerate(ephys_targets):\n",
    "    # filter out genes with zero std dev to avoid divide-by-zero in correlation\n",
    "    variable_genes = log_cpm_df.loc[:, log_cpm_df.std() > 0]\n",
    "\n",
    "    # select top 50 genes correlated with target\n",
    "    corrs = variable_genes.corrwith(ephys_df[target]).abs()\n",
    "    top_genes = corrs.sort_values(ascending=False).head(50).index\n",
    "\n",
    "    X = variable_genes[top_genes]\n",
    "    y = ephys_df[target]\n",
    "\n",
    "    # remove missing values\n",
    "    valid_idx = X.dropna().index.intersection(y.dropna().index)\n",
    "    X = X.loc[valid_idx]\n",
    "    y = y.loc[valid_idx]\n",
    "\n",
    "    # train/test split\n",
    "    X_train, X_test, y_train, y_test = train_test_split(\n",
    "        X, y, test_size=0.3, random_state=42\n",
    "    )\n",
    "\n",
    "    for j, (name, model) in enumerate(models.items()):\n",
    "        pipe = make_pipeline(StandardScaler(), model)\n",
    "\n",
    "        # cross-validation metrics\n",
    "        scores = cross_validate(\n",
    "            pipe, X, y, cv=5, scoring=[\"r2\", \"neg_root_mean_squared_error\"]\n",
    "        )\n",
    "        results.append(\n",
    "            {\n",
    "                \"Model\": name,\n",
    "                \"Ephys Feature\": target,\n",
    "                \"Mean R2\": scores[\"test_r2\"].mean(),\n",
    "                \"Mean RMSE\": -scores[\n",
    "                    \"test_neg_root_mean_squared_error\"\n",
    "                ].mean(),\n",
    "            }\n",
    "        )\n",
    "\n",
    "        # fit & predict on held-out test set\n",
    "        pipe.fit(X_train, y_train)\n",
    "        y_pred = pipe.predict(X_test)\n",
    "\n",
    "        r2 = r2_score(y_test, y_pred)\n",
    "        rmse = np.sqrt(mean_squared_error(y_test, y_pred) ** 0.5)\n",
    "\n",
    "        ax = axes[i, j]\n",
    "        sns.scatterplot(x=y_test, y=y_pred, ax=ax, alpha=0.7)\n",
    "        ax.plot(\n",
    "            [y_test.min(), y_test.max()], [y_test.min(), y_test.max()], \"r--\"\n",
    "        )\n",
    "        ax.set_title(f\"{target}\\n{name}\\nR² = {r2:.2f}, RMSE = {rmse:.2f}\")\n",
    "        ax.set_xlabel(\"True\")\n",
    "        ax.set_ylabel(\"Predicted\")\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.suptitle(\n",
    "    \"Predicted vs. True Ephys Features (All Models)\", fontsize=18, y=1.03\n",
    ")\n",
    "plt.show()\n",
    "\n",
    "# summary of results\n",
    "results_df = pd.DataFrame(results)\n",
    "print(results_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The results show a very poor performance of the models, which implies a very weak predictive power. Either transcriptomic data alone is insufficient to predict the ephys traits, or the input features need some very strong refining, which we were not able to do timewise."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5.3 Biological Interpretability"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 5.3.1 Identify Genes Most Correlated with Electrophysiological Properties\n",
    "\n",
    "=> Correlate all genes with Rheobase (pA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# filter out non-variable genes\n",
    "variable_genes = log_cpm_df.loc[:, log_cpm_df.std() > 0]\n",
    "\n",
    "correlations = variable_genes.corrwith(ephys_df[\"Rheobase (pA)\"])\n",
    "\n",
    "# drop any NaNs that may still result from missing values\n",
    "correlations = correlations.dropna()\n",
    "\n",
    "top_genes_pos = correlations.sort_values(ascending=False).head(50)\n",
    "top_genes_neg = correlations.sort_values(ascending=True).head(50)\n",
    "\n",
    "# combine for enrichment analysis\n",
    "top_genes = pd.concat([top_genes_pos, top_genes_neg]).index.tolist()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 5.3.2 Perform Gene Ontology (GO) or Pathway Enrichment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gp = GProfiler(return_dataframe=True)\n",
    "results = gp.profile(organism=\"mmusculus\", query=top_genes)\n",
    "\n",
    "# view top GO terms\n",
    "results[\n",
    "    [\"name\", \"p_value\", \"term_size\", \"query_size\", \"intersection_size\"]\n",
    "].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# sort by p-value + select top 15 terms\n",
    "top_terms = results.sort_values(\"p_value\").head(15)\n",
    "\n",
    "sns.barplot(\n",
    "    y=top_terms[\"name\"],\n",
    "    x=-np.log10(top_terms[\"p_value\"]),\n",
    "    hue=top_terms[\"name\"],\n",
    "    legend=False,\n",
    "    palette=\"viridis\",\n",
    ")\n",
    "plt.xlabel(\"-log10(p-value)\")\n",
    "plt.title(\"Top GO Terms Enriched in Ephys-Linked Genes\")\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This plot shows the biological pathways most enriched among the top 100 genes most correlated with Rheobase (pA). The results suggest that genes whose expression levels are highly correlated with Rheobase tend to belong to functional categories critical for neuronal communication. This then supports the idea that electrophysiological phenotypes (like firing threshold) are genetically coupled to fundamental neuronal identity."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# References\n",
    "\n",
    "- Harris, K. D., Hochgerner, H., Skene, N. G., Magno, L., Katona, L., Bengtsson Gonzales, C., Somogyi, P., Kessaris, N., Linnarsson, S., & Hjerling-Leffler, J. (2018). Classes and continua of hippocampal CA1 inhibitory neurons revealed by single-cell transcriptomics. *PLoS Biology*, 16(6), e2006387. https://doi.org/10.1371/journal.pbio.2006387\n",
    "\n",
    "- Lause, J., Berens, P., & Kobak, D. (2024). The art of seeing the elephant in the room: 2D embeddings of single-cell data do make sense. *PLoS Computational Biology*, 20(10), e1012403. https://doi.org/10.1371/journal.pcbi.1012403\n",
    "\n",
    "- Loo, L., Simon, J.M., Xing, L. et al. Single-cell transcriptomic analysis of mouse neocortical development. *Nat Commun* 10, 134 (2019). https://doi.org/10.1038/s41467-018-08079-9\n",
    "\n",
    "- Luecken, M. D., & Theis, F. J. (2019). Current best practices in single-cell RNA-seq analysis: a tutorial. *Molecular Systems Biology*, 15(6), e8746. https://doi.org/10.15252/msb.20188746\n",
    "\n",
    "- Tasic, B., Menon, V., Nguyen, T. N., Kim, T. K., Jarsky, T., Yao, Z., ... & Zeng, H. (2016). Adult mouse cortical cell taxonomy revealed by single cell transcriptomics. *Nature neuroscience*, 19(2), 335-346. https://doi.org/10.1038/nn.4216\n",
    "\n",
    "- Tung, PY., Blischak, J., Hsiao, C. et al. Batch effects and the effective design of single-cell gene expression studies. *Sci Rep* 7, 39921 (2017). https://doi.org/10.1038/srep39921"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "python3",
   "language": "python",
   "name": "python3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
